// @angular/compiler-cli/ngcc/src/analysis/migration_host.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as ts from 'typescript';
import { absoluteFromSourceFile } from '../../../src/ngtsc/file_system';
import { TraitState } from '../../../src/ngtsc/transform';
import { isWithinPackage } from './util';
/**
 * The standard implementation of `MigrationHost`, which is created by the `DecorationAnalyzer`.
 */
export class DefaultMigrationHost {
    constructor(reflectionHost, metadata, evaluator, compiler, entryPointPath) {
        this.reflectionHost = reflectionHost;
        this.metadata = metadata;
        this.evaluator = evaluator;
        this.compiler = compiler;
        this.entryPointPath = entryPointPath;
    }
    injectSyntheticDecorator(clazz, decorator, flags) {
        const migratedTraits = this.compiler.injectSyntheticDecorator(clazz, decorator, flags);
        for (const trait of migratedTraits) {
            if ((trait.state === TraitState.Analyzed || trait.state === TraitState.Resolved) &&
                trait.analysisDiagnostics !== null) {
                trait.analysisDiagnostics = trait.analysisDiagnostics.map(diag => createMigrationDiagnostic(diag, clazz, decorator));
            }
            if (trait.state === TraitState.Resolved && trait.resolveDiagnostics !== null) {
                trait.resolveDiagnostics =
                    trait.resolveDiagnostics.map(diag => createMigrationDiagnostic(diag, clazz, decorator));
            }
        }
    }
    getAllDecorators(clazz) {
        return this.compiler.getAllDecorators(clazz);
    }
    isInScope(clazz) {
        return isWithinPackage(this.entryPointPath, absoluteFromSourceFile(clazz.getSourceFile()));
    }
}
/**
 * Creates a diagnostic from another one, containing additional information about the synthetic
 * decorator.
 */
function createMigrationDiagnostic(diagnostic, source, decorator) {
    const clone = Object.assign({}, diagnostic);
    const chain = [{
            messageText: `Occurs for @${decorator.name} decorator inserted by an automatic migration`,
            category: ts.DiagnosticCategory.Message,
            code: 0,
        }];
    if (decorator.args !== null) {
        const args = decorator.args.map(arg => arg.getText()).join(', ');
        chain.push({
            messageText: `@${decorator.name}(${args})`,
            category: ts.DiagnosticCategory.Message,
            code: 0,
        });
    }
    if (typeof clone.messageText === 'string') {
        clone.messageText = {
            messageText: clone.messageText,
            category: diagnostic.category,
            code: diagnostic.code,
            next: chain,
        };
    }
    else {
        if (clone.messageText.next === undefined) {
            clone.messageText.next = chain;
        }
        else {
            clone.messageText.next.push(...chain);
        }
    }
    return clone;
}
