// @angular/compiler-cli/ngcc/src/analysis/ngcc_trait_compiler.mjs
import { NOOP_PERF_RECORDER } from '../../../src/ngtsc/perf';
import { CompilationMode, DtsTransformRegistry, TraitCompiler } from '../../../src/ngtsc/transform';
import { isDefined } from '../utils';
/**
 * Specializes the `TraitCompiler` for ngcc purposes. Mainly, this includes an alternative way of
 * scanning for classes to compile using the reflection host's `findClassSymbols`, together with
 * support to inject synthetic decorators into the compilation for ad-hoc migrations that ngcc
 * performs.
 */
export class NgccTraitCompiler extends TraitCompiler {
    constructor(handlers, ngccReflector) {
        super(handlers, ngccReflector, NOOP_PERF_RECORDER, new NoIncrementalBuild(), 
        /* compileNonExportedClasses */ true, CompilationMode.FULL, new DtsTransformRegistry(), 
        /* semanticDepGraphUpdater */ null);
        this.ngccReflector = ngccReflector;
    }
    get analyzedFiles() {
        return Array.from(this.fileToClasses.keys());
    }
    /**
     * Analyzes the source file in search for classes to process. For any class that is found in the
     * file, a `ClassRecord` is created and the source file is included in the `analyzedFiles` array.
     */
    analyzeFile(sf) {
        const ngccClassSymbols = this.ngccReflector.findClassSymbols(sf);
        for (const classSymbol of ngccClassSymbols) {
            this.analyzeClass(classSymbol.declaration.valueDeclaration, null);
        }
        return undefined;
    }
    /**
     * Associate a new synthesized decorator, which did not appear in the original source, with a
     * given class.
     * @param clazz the class to receive the new decorator.
     * @param decorator the decorator to inject.
     * @param flags optional bitwise flag to influence the compilation of the decorator.
     */
    injectSyntheticDecorator(clazz, decorator, flags) {
        const migratedTraits = this.detectTraits(clazz, [decorator]);
        if (migratedTraits === null) {
            return [];
        }
        for (const trait of migratedTraits) {
            this.analyzeTrait(clazz, trait, flags);
        }
        return migratedTraits;
    }
    /**
     * Returns all decorators that have been recognized for the provided class, including any
     * synthetically injected decorators.
     * @param clazz the declaration for which the decorators are returned.
     */
    getAllDecorators(clazz) {
        const record = this.recordFor(clazz);
        if (record === null) {
            return null;
        }
        return record.traits.map(trait => trait.detected.decorator).filter(isDefined);
    }
}
class NoIncrementalBuild {
    priorAnalysisFor(sf) {
        return null;
    }
    priorTypeCheckingResultsFor() {
        return null;
    }
    recordSuccessfulTypeCheck() { }
}
