// @angular/core/schematics/migrations/missing-injectable/definition_collector.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as ts from 'typescript';
import { getAngularDecorators } from '../../utils/ng_decorators';
import { getPropertyNameText } from '../../utils/typescript/property_name';
/**
 * Visitor that walks through specified TypeScript nodes and collects all
 * found NgModule, Directive or Component definitions.
 */
export class NgDefinitionCollector {
    constructor(typeChecker) {
        this.typeChecker = typeChecker;
        this.resolvedModules = [];
        this.resolvedDirectives = [];
    }
    visitNode(node) {
        if (ts.isClassDeclaration(node)) {
            this.visitClassDeclaration(node);
        }
        ts.forEachChild(node, n => this.visitNode(n));
    }
    visitClassDeclaration(node) {
        if (!node.decorators || !node.decorators.length) {
            return;
        }
        const ngDecorators = getAngularDecorators(this.typeChecker, node.decorators);
        const directiveDecorator = ngDecorators.find(({ name }) => name === 'Component' || name == 'Directive');
        const ngModuleDecorator = ngDecorators.find(({ name }) => name === 'NgModule');
        if (ngModuleDecorator) {
            this._visitNgModuleClass(node, ngModuleDecorator);
        }
        else if (directiveDecorator) {
            this._visitDirectiveClass(node, directiveDecorator);
        }
    }
    _visitDirectiveClass(node, decorator) {
        const decoratorCall = decorator.node.expression;
        const metadata = decoratorCall.arguments[0];
        if (!metadata || !ts.isObjectLiteralExpression(metadata)) {
            return;
        }
        const providersNode = metadata.properties.filter(ts.isPropertyAssignment)
            .find(p => getPropertyNameText(p.name) === 'providers');
        const viewProvidersNode = metadata.properties.filter(ts.isPropertyAssignment)
            .find(p => getPropertyNameText(p.name) === 'viewProviders');
        this.resolvedDirectives.push({
            name: node.name ? node.name.text : 'default',
            node,
            decorator,
            providersExpr: providersNode !== undefined ? providersNode.initializer : null,
            viewProvidersExpr: viewProvidersNode !== undefined ? viewProvidersNode.initializer : null,
        });
    }
    _visitNgModuleClass(node, decorator) {
        const decoratorCall = decorator.node.expression;
        const metadata = decoratorCall.arguments[0];
        if (!metadata || !ts.isObjectLiteralExpression(metadata)) {
            return;
        }
        const providersNode = metadata.properties.filter(ts.isPropertyAssignment)
            .find(p => getPropertyNameText(p.name) === 'providers');
        this.resolvedModules.push({
            name: node.name ? node.name.text : 'default',
            node,
            decorator,
            providersExpr: providersNode !== undefined ? providersNode.initializer : null,
        });
    }
}
