// @angular/compiler-cli/src/ngtsc/reflection/src/host.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as ts from 'typescript';
export const Decorator = {
    nodeForError: (decorator) => {
        if (decorator.node !== null) {
            return decorator.node;
        }
        else {
            // TODO(alxhub): we can't rely on narrowing until TS 3.6 is in g3.
            return decorator.synthesizedFor;
        }
    },
};
export function isDecoratorIdentifier(exp) {
    return ts.isIdentifier(exp) ||
        ts.isPropertyAccessExpression(exp) && ts.isIdentifier(exp.expression) &&
            ts.isIdentifier(exp.name);
}
/**
 * An enumeration of possible kinds of class members.
 */
export var ClassMemberKind;
(function (ClassMemberKind) {
    ClassMemberKind[ClassMemberKind["Constructor"] = 0] = "Constructor";
    ClassMemberKind[ClassMemberKind["Getter"] = 1] = "Getter";
    ClassMemberKind[ClassMemberKind["Setter"] = 2] = "Setter";
    ClassMemberKind[ClassMemberKind["Property"] = 3] = "Property";
    ClassMemberKind[ClassMemberKind["Method"] = 4] = "Method";
})(ClassMemberKind || (ClassMemberKind = {}));
/**
 * Possible declarations of known values, such as built-in objects/functions or TypeScript helpers.
 */
export var KnownDeclaration;
(function (KnownDeclaration) {
    /**
     * Indicates the JavaScript global `Object` class.
     */
    KnownDeclaration[KnownDeclaration["JsGlobalObject"] = 0] = "JsGlobalObject";
    /**
     * Indicates the `__assign` TypeScript helper function.
     */
    KnownDeclaration[KnownDeclaration["TsHelperAssign"] = 1] = "TsHelperAssign";
    /**
     * Indicates the `__spread` TypeScript helper function.
     */
    KnownDeclaration[KnownDeclaration["TsHelperSpread"] = 2] = "TsHelperSpread";
    /**
     * Indicates the `__spreadArrays` TypeScript helper function.
     */
    KnownDeclaration[KnownDeclaration["TsHelperSpreadArrays"] = 3] = "TsHelperSpreadArrays";
    /**
     * Indicates the `__spreadArray` TypeScript helper function.
     */
    KnownDeclaration[KnownDeclaration["TsHelperSpreadArray"] = 4] = "TsHelperSpreadArray";
    /**
     * Indicates the `__read` TypeScript helper function.
     */
    KnownDeclaration[KnownDeclaration["TsHelperRead"] = 5] = "TsHelperRead";
})(KnownDeclaration || (KnownDeclaration = {}));
/**
 * Returns true if the `decl` is a `ConcreteDeclaration` (ie. that its `node` property is a
 * `ts.Declaration`).
 */
export function isConcreteDeclaration(decl) {
    return decl.kind === 0 /* Concrete */;
}
