// @angular/compiler-cli/src/ngtsc/translator/src/import_manager.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as ts from 'typescript';
import { NoopImportRewriter } from '../../imports';
export class ImportManager {
    constructor(rewriter = new NoopImportRewriter(), prefix = 'i') {
        this.rewriter = rewriter;
        this.prefix = prefix;
        this.specifierToIdentifier = new Map();
        this.nextIndex = 0;
    }
    generateNamespaceImport(moduleName) {
        if (!this.specifierToIdentifier.has(moduleName)) {
            this.specifierToIdentifier.set(moduleName, ts.createIdentifier(`${this.prefix}${this.nextIndex++}`));
        }
        return this.specifierToIdentifier.get(moduleName);
    }
    generateNamedImport(moduleName, originalSymbol) {
        // First, rewrite the symbol name.
        const symbol = this.rewriter.rewriteSymbol(originalSymbol, moduleName);
        // Ask the rewriter if this symbol should be imported at all. If not, it can be referenced
        // directly (moduleImport: null).
        if (!this.rewriter.shouldImportSymbol(symbol, moduleName)) {
            // The symbol should be referenced directly.
            return { moduleImport: null, symbol };
        }
        // If not, this symbol will be imported using a generated namespace import.
        const moduleImport = this.generateNamespaceImport(moduleName);
        return { moduleImport, symbol };
    }
    getAllImports(contextPath) {
        const imports = [];
        for (const [originalSpecifier, qualifier] of this.specifierToIdentifier) {
            const specifier = this.rewriter.rewriteSpecifier(originalSpecifier, contextPath);
            imports.push({
                specifier,
                qualifier,
            });
        }
        return imports;
    }
}
