// @angular/compiler-cli/ngcc/src/execution/single_process_executor.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __awaiter } from "tslib";
export class SingleProcessorExecutorBase {
    constructor(logger, createTaskCompletedCallback) {
        this.logger = logger;
        this.createTaskCompletedCallback = createTaskCompletedCallback;
    }
    doExecute(analyzeEntryPoints, createCompileFn) {
        this.logger.debug(`Running ngcc on ${this.constructor.name}.`);
        const taskQueue = analyzeEntryPoints();
        const onTaskCompleted = this.createTaskCompletedCallback(taskQueue);
        const compile = createCompileFn(() => { }, onTaskCompleted);
        // Process all tasks.
        this.logger.debug('Processing tasks...');
        const startTime = Date.now();
        while (!taskQueue.allTasksCompleted) {
            const task = taskQueue.getNextTask();
            compile(task);
            taskQueue.markAsCompleted(task);
        }
        const duration = Math.round((Date.now() - startTime) / 1000);
        this.logger.debug(`Processed tasks in ${duration}s.`);
    }
}
/**
 * An `Executor` that processes all tasks serially and completes synchronously.
 */
export class SingleProcessExecutorSync extends SingleProcessorExecutorBase {
    constructor(logger, lockFile, createTaskCompletedCallback) {
        super(logger, createTaskCompletedCallback);
        this.lockFile = lockFile;
    }
    execute(analyzeEntryPoints, createCompileFn) {
        this.lockFile.lock(() => this.doExecute(analyzeEntryPoints, createCompileFn));
    }
}
/**
 * An `Executor` that processes all tasks serially, but still completes asynchronously.
 */
export class SingleProcessExecutorAsync extends SingleProcessorExecutorBase {
    constructor(logger, lockFile, createTaskCompletedCallback) {
        super(logger, createTaskCompletedCallback);
        this.lockFile = lockFile;
    }
    execute(analyzeEntryPoints, createCompileFn) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.lockFile.lock(() => __awaiter(this, void 0, void 0, function* () { return this.doExecute(analyzeEntryPoints, createCompileFn); }));
        });
    }
}
