// @angular/compiler-cli/linker/babel/src/es2015_linker_plugin.mjs
import { __rest } from "tslib";
import * as t from '@babel/types';
import { FileLinker, isFatalLinkerError, LinkerEnvironment } from '../../../linker';
import { BabelAstFactory } from './ast/babel_ast_factory';
import { BabelAstHost } from './ast/babel_ast_host';
import { BabelDeclarationScope } from './babel_declaration_scope';
/**
 * Create a Babel plugin that visits the program, identifying and linking partial declarations.
 *
 * The plugin delegates most of its work to a generic `FileLinker` for each file (`t.Program` in
 * Babel) that is visited.
 */
export function createEs2015LinkerPlugin(_a) {
    var { fileSystem, logger } = _a, options = __rest(_a, ["fileSystem", "logger"]);
    let fileLinker = null;
    return {
        visitor: {
            Program: {
                /**
                 * Create a new `FileLinker` as we enter each file (`t.Program` in Babel).
                 */
                enter(path) {
                    var _a, _b;
                    assertNull(fileLinker);
                    // Babel can be configured with a `filename` or `relativeFilename` (or both, or neither) -
                    // possibly relative to the optional `cwd` path.
                    const file = path.hub.file;
                    const filename = (_a = file.opts.filename) !== null && _a !== void 0 ? _a : file.opts.filenameRelative;
                    if (!filename) {
                        throw new Error('No filename (nor filenameRelative) provided by Babel. This is required for the linking of partially compiled directives and components.');
                    }
                    const sourceUrl = fileSystem.resolve((_b = file.opts.cwd) !== null && _b !== void 0 ? _b : '.', filename);
                    const linkerEnvironment = LinkerEnvironment.create(fileSystem, logger, new BabelAstHost(), new BabelAstFactory(sourceUrl), options);
                    fileLinker = new FileLinker(linkerEnvironment, sourceUrl, file.code);
                },
                /**
                 * On exiting the file, insert any shared constant statements that were generated during
                 * linking of the partial declarations.
                 */
                exit() {
                    assertNotNull(fileLinker);
                    for (const { constantScope, statements } of fileLinker.getConstantStatements()) {
                        insertStatements(constantScope, statements);
                    }
                    fileLinker = null;
                }
            },
            /**
             * Test each call expression to see if it is a partial declaration; it if is then replace it
             * with the results of linking the declaration.
             */
            CallExpression(call) {
                if (fileLinker === null) {
                    // Any statements that are inserted upon program exit will be visited outside of an active
                    // linker context. These call expressions are known not to contain partial declarations,
                    // so it's safe to skip visiting those call expressions.
                    return;
                }
                try {
                    const calleeName = getCalleeName(call);
                    if (calleeName === null) {
                        return;
                    }
                    const args = call.node.arguments;
                    if (!fileLinker.isPartialDeclaration(calleeName) || !isExpressionArray(args)) {
                        return;
                    }
                    const declarationScope = new BabelDeclarationScope(call.scope);
                    const replacement = fileLinker.linkPartialDeclaration(calleeName, args, declarationScope);
                    call.replaceWith(replacement);
                }
                catch (e) {
                    const node = isFatalLinkerError(e) ? e.node : call.node;
                    throw buildCodeFrameError(call.hub.file, e.message, node);
                }
            }
        }
    };
}
/**
 * Insert the `statements` at the location defined by `path`.
 *
 * The actual insertion strategy depends upon the type of the `path`.
 */
function insertStatements(path, statements) {
    if (path.isFunction()) {
        insertIntoFunction(path, statements);
    }
    else if (path.isProgram()) {
        insertIntoProgram(path, statements);
    }
}
/**
 * Insert the `statements` at the top of the body of the `fn` function.
 */
function insertIntoFunction(fn, statements) {
    const body = fn.get('body');
    body.unshiftContainer('body', statements);
}
/**
 * Insert the `statements` at the top of the `program`, below any import statements.
 */
function insertIntoProgram(program, statements) {
    const body = program.get('body');
    const importStatements = body.filter(statement => statement.isImportDeclaration());
    if (importStatements.length === 0) {
        program.unshiftContainer('body', statements);
    }
    else {
        importStatements[importStatements.length - 1].insertAfter(statements);
    }
}
function getCalleeName(call) {
    const callee = call.node.callee;
    if (t.isIdentifier(callee)) {
        return callee.name;
    }
    else if (t.isMemberExpression(callee) && t.isIdentifier(callee.property)) {
        return callee.property.name;
    }
    else if (t.isMemberExpression(callee) && t.isStringLiteral(callee.property)) {
        return callee.property.value;
    }
    else {
        return null;
    }
}
/**
 * Return true if all the `nodes` are Babel expressions.
 */
function isExpressionArray(nodes) {
    return nodes.every(node => t.isExpression(node));
}
/**
 * Assert that the given `obj` is `null`.
 */
function assertNull(obj) {
    if (obj !== null) {
        throw new Error('BUG - expected `obj` to be null');
    }
}
/**
 * Assert that the given `obj` is not `null`.
 */
function assertNotNull(obj) {
    if (obj === null) {
        throw new Error('BUG - expected `obj` not to be null');
    }
}
/**
 * Create a string representation of an error that includes the code frame of the `node`.
 */
function buildCodeFrameError(file, message, node) {
    const filename = file.opts.filename || '(unknown file)';
    const error = file.buildCodeFrameError(node, message);
    return `${filename}: ${error.message}`;
}
