// @angular/compiler-cli/ngcc/src/execution/tasks/queues/base_task_queue.mjs
import { stringifyTask } from '../utils';
/**
 * A base `TaskQueue` implementation to be used as base for concrete implementations.
 */
export class BaseTaskQueue {
    constructor(logger, tasks, dependencies) {
        this.logger = logger;
        this.tasks = tasks;
        this.dependencies = dependencies;
        this.inProgressTasks = new Set();
        /**
         * A map of tasks that should be skipped, mapped to the task that caused them to be skipped.
         */
        this.tasksToSkip = new Map();
    }
    get allTasksCompleted() {
        return (this.tasks.length === 0) && (this.inProgressTasks.size === 0);
    }
    getNextTask() {
        let nextTask = this.computeNextTask();
        while (nextTask !== null) {
            if (!this.tasksToSkip.has(nextTask)) {
                break;
            }
            // We are skipping this task so mark it as complete
            this.markAsCompleted(nextTask);
            const failedTask = this.tasksToSkip.get(nextTask);
            this.logger.warn(`Skipping processing of ${nextTask.entryPoint.name} because its dependency ${failedTask.entryPoint.name} failed to compile.`);
            nextTask = this.computeNextTask();
        }
        return nextTask;
    }
    markAsCompleted(task) {
        if (!this.inProgressTasks.has(task)) {
            throw new Error(`Trying to mark task that was not in progress as completed: ${stringifyTask(task)}`);
        }
        this.inProgressTasks.delete(task);
    }
    markAsFailed(task) {
        if (this.dependencies.has(task)) {
            for (const dependentTask of this.dependencies.get(task)) {
                this.skipDependentTasks(dependentTask, task);
            }
        }
    }
    markAsUnprocessed(task) {
        if (!this.inProgressTasks.has(task)) {
            throw new Error(`Trying to mark task that was not in progress as unprocessed: ${stringifyTask(task)}`);
        }
        this.inProgressTasks.delete(task);
        this.tasks.unshift(task);
    }
    toString() {
        const inProgTasks = Array.from(this.inProgressTasks);
        return `${this.constructor.name}\n` +
            `  All tasks completed: ${this.allTasksCompleted}\n` +
            `  Unprocessed tasks (${this.tasks.length}): ${this.stringifyTasks(this.tasks, '    ')}\n` +
            `  In-progress tasks (${inProgTasks.length}): ${this.stringifyTasks(inProgTasks, '    ')}`;
    }
    /**
     * Mark the given `task` as to be skipped, then recursive skip all its dependents.
     *
     * @param task The task to skip
     * @param failedTask The task that failed, causing this task to be skipped
     */
    skipDependentTasks(task, failedTask) {
        this.tasksToSkip.set(task, failedTask);
        if (this.dependencies.has(task)) {
            for (const dependentTask of this.dependencies.get(task)) {
                this.skipDependentTasks(dependentTask, failedTask);
            }
        }
    }
    stringifyTasks(tasks, indentation) {
        return tasks.map(task => `\n${indentation}- ${stringifyTask(task)}`).join('');
    }
}
