// @angular/compiler-cli/linker/src/linker_import_generator.mjs
import { FatalLinkerError } from './fatal_linker_error';
/**
 * A class that is used to generate imports when translating from Angular Output AST to an AST to
 * render, such as Babel.
 *
 * Note that, in the linker, there can only be imports from `@angular/core` and that these imports
 * must be achieved by property access on an `ng` namespace identifer, which is passed in via the
 * constructor.
 */
export class LinkerImportGenerator {
    constructor(ngImport) {
        this.ngImport = ngImport;
    }
    generateNamespaceImport(moduleName) {
        this.assertModuleName(moduleName);
        return this.ngImport;
    }
    generateNamedImport(moduleName, originalSymbol) {
        this.assertModuleName(moduleName);
        return { moduleImport: this.ngImport, symbol: originalSymbol };
    }
    assertModuleName(moduleName) {
        if (moduleName !== '@angular/core') {
            throw new FatalLinkerError(this.ngImport, `Unable to import from anything other than '@angular/core'`);
        }
    }
}
