// @angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_pipe_linker_1.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { compilePipeFromMetadata } from '@angular/compiler';
import { FatalLinkerError } from '../../fatal_linker_error';
import { wrapReference } from './util';
/**
 * A `PartialLinker` that is designed to process `ɵɵngDeclarePipe()` call expressions.
 */
export class PartialPipeLinkerVersion1 {
    constructor() { }
    linkPartialDeclaration(constantPool, metaObj) {
        const meta = toR3PipeMeta(metaObj);
        const def = compilePipeFromMetadata(meta);
        return def.expression;
    }
}
/**
 * Derives the `R3PipeMetadata` structure from the AST object.
 */
export function toR3PipeMeta(metaObj) {
    const typeExpr = metaObj.getValue('type');
    const typeName = typeExpr.getSymbolName();
    if (typeName === null) {
        throw new FatalLinkerError(typeExpr.expression, 'Unsupported type, its name could not be determined');
    }
    const pure = metaObj.has('pure') ? metaObj.getBoolean('pure') : true;
    return {
        name: typeName,
        type: wrapReference(typeExpr.getOpaque()),
        internalType: metaObj.getOpaque('type'),
        typeArgumentCount: 0,
        deps: null,
        pipeName: metaObj.getString('name'),
        pure,
    };
}
