// @angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_factory_linker_1.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { compileFactoryFunction, FactoryTarget } from '@angular/compiler';
import { FatalLinkerError } from '../../fatal_linker_error';
import { getDependency, parseEnum, wrapReference } from './util';
/**
 * A `PartialLinker` that is designed to process `ɵɵngDeclareFactory()` call expressions.
 */
export class PartialFactoryLinkerVersion1 {
    linkPartialDeclaration(constantPool, metaObj) {
        const meta = toR3FactoryMeta(metaObj);
        const def = compileFactoryFunction(meta);
        return def.expression;
    }
}
/**
 * Derives the `R3FactoryMetadata` structure from the AST object.
 */
export function toR3FactoryMeta(metaObj) {
    const typeExpr = metaObj.getValue('type');
    const typeName = typeExpr.getSymbolName();
    if (typeName === null) {
        throw new FatalLinkerError(typeExpr.expression, 'Unsupported type, its name could not be determined');
    }
    return {
        name: typeName,
        type: wrapReference(typeExpr.getOpaque()),
        internalType: metaObj.getOpaque('type'),
        typeArgumentCount: 0,
        target: parseEnum(metaObj.getValue('target'), FactoryTarget),
        deps: getDependencies(metaObj, 'deps'),
    };
}
function getDependencies(metaObj, propName) {
    if (!metaObj.has(propName)) {
        return null;
    }
    const deps = metaObj.getValue(propName);
    if (deps.isArray()) {
        return deps.getArray().map(dep => getDependency(dep.getObject()));
    }
    if (deps.isString()) {
        return 'invalid';
    }
    return null;
}
