// @angular/compiler-cli/linker/src/file_linker/partial_linkers/partial_injectable_linker_1.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { compileInjectable, createR3ProviderExpression } from '@angular/compiler';
import * as o from '@angular/compiler/src/output/output_ast';
import { FatalLinkerError } from '../../fatal_linker_error';
import { extractForwardRef, getDependency, wrapReference } from './util';
/**
 * A `PartialLinker` that is designed to process `ɵɵngDeclareInjectable()` call expressions.
 */
export class PartialInjectableLinkerVersion1 {
    linkPartialDeclaration(constantPool, metaObj) {
        const meta = toR3InjectableMeta(metaObj);
        const def = compileInjectable(meta, /* resolveForwardRefs */ false);
        return def.expression;
    }
}
/**
 * Derives the `R3InjectableMetadata` structure from the AST object.
 */
export function toR3InjectableMeta(metaObj) {
    const typeExpr = metaObj.getValue('type');
    const typeName = typeExpr.getSymbolName();
    if (typeName === null) {
        throw new FatalLinkerError(typeExpr.expression, 'Unsupported type, its name could not be determined');
    }
    const meta = {
        name: typeName,
        type: wrapReference(typeExpr.getOpaque()),
        internalType: typeExpr.getOpaque(),
        typeArgumentCount: 0,
        providedIn: metaObj.has('providedIn') ? extractForwardRef(metaObj.getValue('providedIn')) :
            createR3ProviderExpression(o.literal(null), false),
    };
    if (metaObj.has('useClass')) {
        meta.useClass = extractForwardRef(metaObj.getValue('useClass'));
    }
    if (metaObj.has('useFactory')) {
        meta.useFactory = metaObj.getOpaque('useFactory');
    }
    if (metaObj.has('useExisting')) {
        meta.useExisting = extractForwardRef(metaObj.getValue('useExisting'));
    }
    if (metaObj.has('useValue')) {
        meta.useValue = extractForwardRef(metaObj.getValue('useValue'));
    }
    if (metaObj.has('deps')) {
        meta.deps = metaObj.getArray('deps').map(dep => getDependency(dep.getObject()));
    }
    return meta;
}
