// @angular/compiler-cli/src/ngtsc/typecheck/src/checker.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { CssSelector, DomElementSchemaRegistry } from '@angular/compiler';
import { absoluteFrom, absoluteFromSourceFile, getSourceFileOrError } from '../../file_system';
import { Reference } from '../../imports';
import { PerfCheckpoint, PerfEvent, PerfPhase } from '../../perf';
import { UpdateMode } from '../../program_driver';
import { isNamedClassDeclaration } from '../../reflection';
import { isShim } from '../../shims';
import { getSourceFileOrNull, isSymbolWithValueDeclaration } from '../../util/src/typescript';
import { OptimizeFor } from '../api';
import { CompletionEngine } from './completion';
import { InliningMode, TypeCheckContextImpl } from './context';
import { shouldReportDiagnostic, translateDiagnostic } from './diagnostics';
import { TypeCheckShimGenerator } from './shim';
import { TemplateSourceManager } from './source';
import { findTypeCheckBlock, getTemplateMapping } from './tcb_util';
import { SymbolBuilder } from './template_symbol_builder';
const REGISTRY = new DomElementSchemaRegistry();
/**
 * Primary template type-checking engine, which performs type-checking using a
 * `TypeCheckingProgramStrategy` for type-checking program maintenance, and the
 * `ProgramTypeCheckAdapter` for generation of template type-checking code.
 */
export class TemplateTypeCheckerImpl {
    constructor(originalProgram, programDriver, typeCheckAdapter, config, refEmitter, reflector, compilerHost, priorBuild, componentScopeReader, typeCheckScopeRegistry, perf) {
        this.originalProgram = originalProgram;
        this.programDriver = programDriver;
        this.typeCheckAdapter = typeCheckAdapter;
        this.config = config;
        this.refEmitter = refEmitter;
        this.reflector = reflector;
        this.compilerHost = compilerHost;
        this.priorBuild = priorBuild;
        this.componentScopeReader = componentScopeReader;
        this.typeCheckScopeRegistry = typeCheckScopeRegistry;
        this.perf = perf;
        this.state = new Map();
        /**
         * Stores the `CompletionEngine` which powers autocompletion for each component class.
         *
         * Must be invalidated whenever the component's template or the `ts.Program` changes. Invalidation
         * on template changes is performed within this `TemplateTypeCheckerImpl` instance. When the
         * `ts.Program` changes, the `TemplateTypeCheckerImpl` as a whole is destroyed and replaced.
         */
        this.completionCache = new Map();
        /**
         * Stores the `SymbolBuilder` which creates symbols for each component class.
         *
         * Must be invalidated whenever the component's template or the `ts.Program` changes. Invalidation
         * on template changes is performed within this `TemplateTypeCheckerImpl` instance. When the
         * `ts.Program` changes, the `TemplateTypeCheckerImpl` as a whole is destroyed and replaced.
         */
        this.symbolBuilderCache = new Map();
        /**
         * Stores directives and pipes that are in scope for each component.
         *
         * Unlike other caches, the scope of a component is not affected by its template. It will be
         * destroyed when the `ts.Program` changes and the `TemplateTypeCheckerImpl` as a whole is
         * destroyed and replaced.
         */
        this.scopeCache = new Map();
        /**
         * Stores potential element tags for each component (a union of DOM tags as well as directive
         * tags).
         *
         * Unlike other caches, the scope of a component is not affected by its template. It will be
         * destroyed when the `ts.Program` changes and the `TemplateTypeCheckerImpl` as a whole is
         * destroyed and replaced.
         */
        this.elementTagCache = new Map();
        this.isComplete = false;
    }
    getTemplate(component) {
        const { data } = this.getLatestComponentState(component);
        if (data === null) {
            return null;
        }
        return data.template;
    }
    getLatestComponentState(component) {
        this.ensureShimForComponent(component);
        const sf = component.getSourceFile();
        const sfPath = absoluteFromSourceFile(sf);
        const shimPath = TypeCheckShimGenerator.shimFor(sfPath);
        const fileRecord = this.getFileData(sfPath);
        if (!fileRecord.shimData.has(shimPath)) {
            return { data: null, tcb: null, shimPath };
        }
        const templateId = fileRecord.sourceManager.getTemplateId(component);
        const shimRecord = fileRecord.shimData.get(shimPath);
        const id = fileRecord.sourceManager.getTemplateId(component);
        const program = this.programDriver.getProgram();
        const shimSf = getSourceFileOrNull(program, shimPath);
        if (shimSf === null || !fileRecord.shimData.has(shimPath)) {
            throw new Error(`Error: no shim file in program: ${shimPath}`);
        }
        let tcb = findTypeCheckBlock(shimSf, id, /*isDiagnosticsRequest*/ false);
        if (tcb === null) {
            // Try for an inline block.
            const inlineSf = getSourceFileOrError(program, sfPath);
            tcb = findTypeCheckBlock(inlineSf, id, /*isDiagnosticsRequest*/ false);
        }
        let data = null;
        if (shimRecord.templates.has(templateId)) {
            data = shimRecord.templates.get(templateId);
        }
        return { data, tcb, shimPath };
    }
    isTrackedTypeCheckFile(filePath) {
        return this.getFileAndShimRecordsForPath(filePath) !== null;
    }
    getFileAndShimRecordsForPath(shimPath) {
        for (const fileRecord of this.state.values()) {
            if (fileRecord.shimData.has(shimPath)) {
                return { fileRecord, shimRecord: fileRecord.shimData.get(shimPath) };
            }
        }
        return null;
    }
    getTemplateMappingAtShimLocation({ shimPath, positionInShimFile }) {
        const records = this.getFileAndShimRecordsForPath(absoluteFrom(shimPath));
        if (records === null) {
            return null;
        }
        const { fileRecord } = records;
        const shimSf = this.programDriver.getProgram().getSourceFile(absoluteFrom(shimPath));
        if (shimSf === undefined) {
            return null;
        }
        return getTemplateMapping(shimSf, positionInShimFile, fileRecord.sourceManager, /*isDiagnosticsRequest*/ false);
    }
    generateAllTypeCheckBlocks() {
        this.ensureAllShimsForAllFiles();
    }
    /**
     * Retrieve type-checking and template parse diagnostics from the given `ts.SourceFile` using the
     * most recent type-checking program.
     */
    getDiagnosticsForFile(sf, optimizeFor) {
        switch (optimizeFor) {
            case OptimizeFor.WholeProgram:
                this.ensureAllShimsForAllFiles();
                break;
            case OptimizeFor.SingleFile:
                this.ensureAllShimsForOneFile(sf);
                break;
        }
        return this.perf.inPhase(PerfPhase.TtcDiagnostics, () => {
            const sfPath = absoluteFromSourceFile(sf);
            const fileRecord = this.state.get(sfPath);
            const typeCheckProgram = this.programDriver.getProgram();
            const diagnostics = [];
            if (fileRecord.hasInlines) {
                const inlineSf = getSourceFileOrError(typeCheckProgram, sfPath);
                diagnostics.push(...typeCheckProgram.getSemanticDiagnostics(inlineSf).map(diag => convertDiagnostic(diag, fileRecord.sourceManager)));
            }
            for (const [shimPath, shimRecord] of fileRecord.shimData) {
                const shimSf = getSourceFileOrError(typeCheckProgram, shimPath);
                diagnostics.push(...typeCheckProgram.getSemanticDiagnostics(shimSf).map(diag => convertDiagnostic(diag, fileRecord.sourceManager)));
                diagnostics.push(...shimRecord.genesisDiagnostics);
                for (const templateData of shimRecord.templates.values()) {
                    diagnostics.push(...templateData.templateDiagnostics);
                }
            }
            return diagnostics.filter((diag) => diag !== null);
        });
    }
    getDiagnosticsForComponent(component) {
        this.ensureShimForComponent(component);
        return this.perf.inPhase(PerfPhase.TtcDiagnostics, () => {
            const sf = component.getSourceFile();
            const sfPath = absoluteFromSourceFile(sf);
            const shimPath = TypeCheckShimGenerator.shimFor(sfPath);
            const fileRecord = this.getFileData(sfPath);
            if (!fileRecord.shimData.has(shimPath)) {
                return [];
            }
            const templateId = fileRecord.sourceManager.getTemplateId(component);
            const shimRecord = fileRecord.shimData.get(shimPath);
            const typeCheckProgram = this.programDriver.getProgram();
            const diagnostics = [];
            if (shimRecord.hasInlines) {
                const inlineSf = getSourceFileOrError(typeCheckProgram, sfPath);
                diagnostics.push(...typeCheckProgram.getSemanticDiagnostics(inlineSf).map(diag => convertDiagnostic(diag, fileRecord.sourceManager)));
            }
            const shimSf = getSourceFileOrError(typeCheckProgram, shimPath);
            diagnostics.push(...typeCheckProgram.getSemanticDiagnostics(shimSf).map(diag => convertDiagnostic(diag, fileRecord.sourceManager)));
            diagnostics.push(...shimRecord.genesisDiagnostics);
            for (const templateData of shimRecord.templates.values()) {
                diagnostics.push(...templateData.templateDiagnostics);
            }
            return diagnostics.filter((diag) => diag !== null && diag.templateId === templateId);
        });
    }
    getTypeCheckBlock(component) {
        return this.getLatestComponentState(component).tcb;
    }
    getGlobalCompletions(context, component, node) {
        const engine = this.getOrCreateCompletionEngine(component);
        if (engine === null) {
            return null;
        }
        return this.perf.inPhase(PerfPhase.TtcAutocompletion, () => engine.getGlobalCompletions(context, node));
    }
    getExpressionCompletionLocation(ast, component) {
        const engine = this.getOrCreateCompletionEngine(component);
        if (engine === null) {
            return null;
        }
        return this.perf.inPhase(PerfPhase.TtcAutocompletion, () => engine.getExpressionCompletionLocation(ast));
    }
    invalidateClass(clazz) {
        this.completionCache.delete(clazz);
        this.symbolBuilderCache.delete(clazz);
        this.scopeCache.delete(clazz);
        this.elementTagCache.delete(clazz);
        const sf = clazz.getSourceFile();
        const sfPath = absoluteFromSourceFile(sf);
        const shimPath = TypeCheckShimGenerator.shimFor(sfPath);
        const fileData = this.getFileData(sfPath);
        const templateId = fileData.sourceManager.getTemplateId(clazz);
        fileData.shimData.delete(shimPath);
        fileData.isComplete = false;
        this.isComplete = false;
    }
    getOrCreateCompletionEngine(component) {
        if (this.completionCache.has(component)) {
            return this.completionCache.get(component);
        }
        const { tcb, data, shimPath } = this.getLatestComponentState(component);
        if (tcb === null || data === null) {
            return null;
        }
        const engine = new CompletionEngine(tcb, data, shimPath);
        this.completionCache.set(component, engine);
        return engine;
    }
    maybeAdoptPriorResultsForFile(sf) {
        const sfPath = absoluteFromSourceFile(sf);
        if (this.state.has(sfPath)) {
            const existingResults = this.state.get(sfPath);
            if (existingResults.isComplete) {
                // All data for this file has already been generated, so no need to adopt anything.
                return;
            }
        }
        const previousResults = this.priorBuild.priorTypeCheckingResultsFor(sf);
        if (previousResults === null || !previousResults.isComplete) {
            return;
        }
        this.perf.eventCount(PerfEvent.ReuseTypeCheckFile);
        this.state.set(sfPath, previousResults);
    }
    ensureAllShimsForAllFiles() {
        if (this.isComplete) {
            return;
        }
        this.perf.inPhase(PerfPhase.TcbGeneration, () => {
            const host = new WholeProgramTypeCheckingHost(this);
            const ctx = this.newContext(host);
            for (const sf of this.originalProgram.getSourceFiles()) {
                if (sf.isDeclarationFile || isShim(sf)) {
                    continue;
                }
                this.maybeAdoptPriorResultsForFile(sf);
                const sfPath = absoluteFromSourceFile(sf);
                const fileData = this.getFileData(sfPath);
                if (fileData.isComplete) {
                    continue;
                }
                this.typeCheckAdapter.typeCheck(sf, ctx);
                fileData.isComplete = true;
            }
            this.updateFromContext(ctx);
            this.isComplete = true;
        });
    }
    ensureAllShimsForOneFile(sf) {
        this.perf.inPhase(PerfPhase.TcbGeneration, () => {
            this.maybeAdoptPriorResultsForFile(sf);
            const sfPath = absoluteFromSourceFile(sf);
            const fileData = this.getFileData(sfPath);
            if (fileData.isComplete) {
                // All data for this file is present and accounted for already.
                return;
            }
            const host = new SingleFileTypeCheckingHost(sfPath, fileData, this);
            const ctx = this.newContext(host);
            this.typeCheckAdapter.typeCheck(sf, ctx);
            fileData.isComplete = true;
            this.updateFromContext(ctx);
        });
    }
    ensureShimForComponent(component) {
        const sf = component.getSourceFile();
        const sfPath = absoluteFromSourceFile(sf);
        const shimPath = TypeCheckShimGenerator.shimFor(sfPath);
        this.maybeAdoptPriorResultsForFile(sf);
        const fileData = this.getFileData(sfPath);
        if (fileData.shimData.has(shimPath)) {
            // All data for this component is available.
            return;
        }
        const host = new SingleShimTypeCheckingHost(sfPath, fileData, this, shimPath);
        const ctx = this.newContext(host);
        this.typeCheckAdapter.typeCheck(sf, ctx);
        this.updateFromContext(ctx);
    }
    newContext(host) {
        const inlining = this.programDriver.supportsInlineOperations ? InliningMode.InlineOps : InliningMode.Error;
        return new TypeCheckContextImpl(this.config, this.compilerHost, this.refEmitter, this.reflector, host, inlining, this.perf);
    }
    /**
     * Remove any shim data that depends on inline operations applied to the type-checking program.
     *
     * This can be useful if new inlines need to be applied, and it's not possible to guarantee that
     * they won't overwrite or corrupt existing inlines that are used by such shims.
     */
    clearAllShimDataUsingInlines() {
        for (const fileData of this.state.values()) {
            if (!fileData.hasInlines) {
                continue;
            }
            for (const [shimFile, shimData] of fileData.shimData.entries()) {
                if (shimData.hasInlines) {
                    fileData.shimData.delete(shimFile);
                }
            }
            fileData.hasInlines = false;
            fileData.isComplete = false;
            this.isComplete = false;
        }
    }
    updateFromContext(ctx) {
        const updates = ctx.finalize();
        return this.perf.inPhase(PerfPhase.TcbUpdateProgram, () => {
            if (updates.size > 0) {
                this.perf.eventCount(PerfEvent.UpdateTypeCheckProgram);
            }
            this.programDriver.updateFiles(updates, UpdateMode.Incremental);
            this.priorBuild.recordSuccessfulTypeCheck(this.state);
            this.perf.memory(PerfCheckpoint.TtcUpdateProgram);
        });
    }
    getFileData(path) {
        if (!this.state.has(path)) {
            this.state.set(path, {
                hasInlines: false,
                sourceManager: new TemplateSourceManager(),
                isComplete: false,
                shimData: new Map(),
            });
        }
        return this.state.get(path);
    }
    getSymbolOfNode(node, component) {
        const builder = this.getOrCreateSymbolBuilder(component);
        if (builder === null) {
            return null;
        }
        return this.perf.inPhase(PerfPhase.TtcSymbol, () => builder.getSymbol(node));
    }
    getOrCreateSymbolBuilder(component) {
        if (this.symbolBuilderCache.has(component)) {
            return this.symbolBuilderCache.get(component);
        }
        const { tcb, data, shimPath } = this.getLatestComponentState(component);
        if (tcb === null || data === null) {
            return null;
        }
        const builder = new SymbolBuilder(shimPath, tcb, data, this.componentScopeReader, () => this.programDriver.getProgram().getTypeChecker());
        this.symbolBuilderCache.set(component, builder);
        return builder;
    }
    getDirectivesInScope(component) {
        const data = this.getScopeData(component);
        if (data === null) {
            return null;
        }
        return data.directives;
    }
    getPipesInScope(component) {
        const data = this.getScopeData(component);
        if (data === null) {
            return null;
        }
        return data.pipes;
    }
    getDirectiveMetadata(dir) {
        if (!isNamedClassDeclaration(dir)) {
            return null;
        }
        return this.typeCheckScopeRegistry.getTypeCheckDirectiveMetadata(new Reference(dir));
    }
    getPotentialElementTags(component) {
        if (this.elementTagCache.has(component)) {
            return this.elementTagCache.get(component);
        }
        const tagMap = new Map();
        for (const tag of REGISTRY.allKnownElementNames()) {
            tagMap.set(tag, null);
        }
        const scope = this.getScopeData(component);
        if (scope !== null) {
            for (const directive of scope.directives) {
                for (const selector of CssSelector.parse(directive.selector)) {
                    if (selector.element === null || tagMap.has(selector.element)) {
                        // Skip this directive if it doesn't match an element tag, or if another directive has
                        // already been included with the same element name.
                        continue;
                    }
                    tagMap.set(selector.element, directive);
                }
            }
        }
        this.elementTagCache.set(component, tagMap);
        return tagMap;
    }
    getPotentialDomBindings(tagName) {
        const attributes = REGISTRY.allKnownAttributesOfElement(tagName);
        return attributes.map(attribute => ({
            attribute,
            property: REGISTRY.getMappedPropName(attribute),
        }));
    }
    getScopeData(component) {
        if (this.scopeCache.has(component)) {
            return this.scopeCache.get(component);
        }
        if (!isNamedClassDeclaration(component)) {
            throw new Error(`AssertionError: components must have names`);
        }
        const scope = this.componentScopeReader.getScopeForComponent(component);
        if (scope === null) {
            return null;
        }
        const data = {
            directives: [],
            pipes: [],
            isPoisoned: scope.compilation.isPoisoned,
        };
        const typeChecker = this.programDriver.getProgram().getTypeChecker();
        for (const dir of scope.compilation.directives) {
            if (dir.selector === null) {
                // Skip this directive, it can't be added to a template anyway.
                continue;
            }
            const tsSymbol = typeChecker.getSymbolAtLocation(dir.ref.node.name);
            if (!isSymbolWithValueDeclaration(tsSymbol)) {
                continue;
            }
            let ngModule = null;
            const moduleScopeOfDir = this.componentScopeReader.getScopeForComponent(dir.ref.node);
            if (moduleScopeOfDir !== null) {
                ngModule = moduleScopeOfDir.ngModule;
            }
            data.directives.push({
                isComponent: dir.isComponent,
                isStructural: dir.isStructural,
                selector: dir.selector,
                tsSymbol,
                ngModule,
            });
        }
        for (const pipe of scope.compilation.pipes) {
            const tsSymbol = typeChecker.getSymbolAtLocation(pipe.ref.node.name);
            if (tsSymbol === undefined) {
                continue;
            }
            data.pipes.push({
                name: pipe.name,
                tsSymbol,
            });
        }
        this.scopeCache.set(component, data);
        return data;
    }
}
function convertDiagnostic(diag, sourceResolver) {
    if (!shouldReportDiagnostic(diag)) {
        return null;
    }
    return translateDiagnostic(diag, sourceResolver);
}
/**
 * Drives a `TypeCheckContext` to generate type-checking code for every component in the program.
 */
class WholeProgramTypeCheckingHost {
    constructor(impl) {
        this.impl = impl;
    }
    getSourceManager(sfPath) {
        return this.impl.getFileData(sfPath).sourceManager;
    }
    shouldCheckComponent(node) {
        const sfPath = absoluteFromSourceFile(node.getSourceFile());
        const shimPath = TypeCheckShimGenerator.shimFor(sfPath);
        const fileData = this.impl.getFileData(sfPath);
        // The component needs to be checked unless the shim which would contain it already exists.
        return !fileData.shimData.has(shimPath);
    }
    recordShimData(sfPath, data) {
        const fileData = this.impl.getFileData(sfPath);
        fileData.shimData.set(data.path, data);
        if (data.hasInlines) {
            fileData.hasInlines = true;
        }
    }
    recordComplete(sfPath) {
        this.impl.getFileData(sfPath).isComplete = true;
    }
}
/**
 * Drives a `TypeCheckContext` to generate type-checking code efficiently for a single input file.
 */
class SingleFileTypeCheckingHost {
    constructor(sfPath, fileData, impl) {
        this.sfPath = sfPath;
        this.fileData = fileData;
        this.impl = impl;
        this.seenInlines = false;
    }
    assertPath(sfPath) {
        if (this.sfPath !== sfPath) {
            throw new Error(`AssertionError: querying TypeCheckingHost outside of assigned file`);
        }
    }
    getSourceManager(sfPath) {
        this.assertPath(sfPath);
        return this.fileData.sourceManager;
    }
    shouldCheckComponent(node) {
        if (this.sfPath !== absoluteFromSourceFile(node.getSourceFile())) {
            return false;
        }
        const shimPath = TypeCheckShimGenerator.shimFor(this.sfPath);
        // Only need to generate a TCB for the class if no shim exists for it currently.
        return !this.fileData.shimData.has(shimPath);
    }
    recordShimData(sfPath, data) {
        this.assertPath(sfPath);
        // Previous type-checking state may have required the use of inlines (assuming they were
        // supported). If the current operation also requires inlines, this presents a problem:
        // generating new inlines may invalidate any old inlines that old state depends on.
        //
        // Rather than resolve this issue by tracking specific dependencies on inlines, if the new state
        // relies on inlines, any old state that relied on them is simply cleared. This happens when the
        // first new state that uses inlines is encountered.
        if (data.hasInlines && !this.seenInlines) {
            this.impl.clearAllShimDataUsingInlines();
            this.seenInlines = true;
        }
        this.fileData.shimData.set(data.path, data);
        if (data.hasInlines) {
            this.fileData.hasInlines = true;
        }
    }
    recordComplete(sfPath) {
        this.assertPath(sfPath);
        this.fileData.isComplete = true;
    }
}
/**
 * Drives a `TypeCheckContext` to generate type-checking code efficiently for only those components
 * which map to a single shim of a single input file.
 */
class SingleShimTypeCheckingHost extends SingleFileTypeCheckingHost {
    constructor(sfPath, fileData, impl, shimPath) {
        super(sfPath, fileData, impl);
        this.shimPath = shimPath;
    }
    shouldCheckNode(node) {
        if (this.sfPath !== absoluteFromSourceFile(node.getSourceFile())) {
            return false;
        }
        // Only generate a TCB for the component if it maps to the requested shim file.
        const shimPath = TypeCheckShimGenerator.shimFor(this.sfPath);
        if (shimPath !== this.shimPath) {
            return false;
        }
        // Only need to generate a TCB for the class if no shim exists for it currently.
        return !this.fileData.shimData.has(shimPath);
    }
}
