// @angular/compiler-cli/src/ngtsc/typecheck/src/template_semantics.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { ImplicitReceiver, RecursiveAstVisitor, TmplAstVariable } from '@angular/compiler';
/**
 * Visits a template and records any semantic errors within its expressions.
 */
export class ExpressionSemanticVisitor extends RecursiveAstVisitor {
    constructor(templateId, boundTarget, oob) {
        super();
        this.templateId = templateId;
        this.boundTarget = boundTarget;
        this.oob = oob;
    }
    visitPropertyWrite(ast, context) {
        super.visitPropertyWrite(ast, context);
        if (!(ast.receiver instanceof ImplicitReceiver)) {
            return;
        }
        const target = this.boundTarget.getExpressionTarget(ast);
        if (target instanceof TmplAstVariable) {
            // Template variables are read-only.
            this.oob.illegalAssignmentToTemplateVar(this.templateId, ast, target);
        }
    }
    static visit(ast, id, boundTarget, oob) {
        ast.visit(new ExpressionSemanticVisitor(id, boundTarget, oob));
    }
}
