// ngx-panzoom/bundles/ngx-panzoom.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('ngx-panzoom', ['exports', '@angular/core', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ngx-panzoom'] = {}, global.ng.core, global.rxjs));
}(this, (function (exports, core, rxjs) { 'use strict';

    var PanZoomComponent = /** @class */ (function () {
        function PanZoomComponent(zone) {
            var _this = this;
            this.zone = zone;
            this.isDragging = false;
            this.lastTick = 0;
            this.isChrome = false;
            this.willChangeNextFrame = true; // used for scaling in Chrome
            this.isMobile = false;
            this.isFirstSync = true;
            this.zoomLevelIsChanging = false;
            this.dragFinishing = false;
            //////////////////////////// END OF LIFECYCLE HOOKS ////////////////////////////
            ////////////////////////////////////////////////////////////////////////////////
            ///////////////////////////////// EVENT HANDLERS ///////////////////////////////
            ////////////////////////////////////////////////////////////////////////////////
            this.onMouseWheel = function (event) {
                // console.log('PanZoomComponent: OnMouseWheel() event:', event);
                if (!(event && event.pageX && event.pageY)) {
                    return;
                }
                var deltaY = event.deltaY;
                if (_this.config.zoomOnMouseWheel) {
                    if (_this.animationParams) {
                        return; // already zooming
                    }
                    if (!_this.config.invertMouseWheel) {
                        deltaY = -deltaY;
                    }
                    var frameElementOffset = _this.getFrameElementOffset();
                    var clickPoint = {
                        x: event.pageX - frameElementOffset.left,
                        y: event.pageY - frameElementOffset.top
                    };
                    _this.lastClickPoint = clickPoint;
                    if (_this.config.freeMouseWheel) {
                        // free wheel scroll
                        _this.freeZoom(clickPoint, deltaY);
                    }
                    else {
                        if (deltaY < 0) {
                            _this.zoomInToPoint(clickPoint);
                        }
                        else if (deltaY > 0) {
                            _this.zoomOutFromPoint(clickPoint);
                        }
                    }
                }
            };
            this.onMousedown = function (event) {
                // console.log('PanZoomComponent: onMousedown()', event);
                /* Check if clicked location is inside element from which
                   dragging is prevented. */
                if (_this.config.noDragFromElementClass
                    && _this.isParentElement(_this.config.noDragFromElementClass, event.srcElement)) {
                    return;
                }
                if (event.button === _this.dragMouseButton || event.type === 'touchstart') {
                    if (event.type !== 'touchstart') {
                        event.preventDefault();
                    }
                    _this.dragFinishing = false;
                    _this.panVelocity = undefined;
                    if (_this.config.panOnClickDrag) {
                        _this.previousPosition = {
                            x: event.pageX,
                            y: event.pageY
                        };
                        _this.lastMouseEventTime = event.timeStamp;
                        _this.isDragging = true;
                        _this.model.isPanning = false;
                        if (_this.isMobile) {
                            _this.zone.runOutsideAngular(function () { return document.addEventListener('touchend', _this.onTouchEnd, false); }); // leave this on document
                            _this.zone.runOutsideAngular(function () { return document.addEventListener('touchmove', _this.onTouchMove, { passive: true, capture: false }); }); // leave this on document
                        }
                        else {
                            _this.zone.runOutsideAngular(function () { return document.addEventListener('mousemove', _this.onMouseMove, { passive: true, capture: false }); }); // leave this on document
                            _this.zone.runOutsideAngular(function () { return document.addEventListener('mouseup', _this.onMouseUp); }); // leave this on document
                        }
                    }
                    return false;
                }
            };
            this.onTouchStart = function (event) {
                // console.log('PanZoomComponent: onTouchStart()', event);
                // console.log('PanZoomComponent: onTouchStart(): touches:', event.touches.length);
                //event.preventDefault();
                if (event.touches.length !== 1) {
                    // multiple touches, get ready for zooming
                    // Calculate x and y distance between touch events
                    var x = event.touches[0].pageX - event.touches[1].pageX;
                    var y = event.touches[0].pageY - event.touches[1].pageY;
                    // Calculate length between touch points with pythagoras
                    // There is no reason to use Math.pow and Math.sqrt as we
                    // only want a relative length and not the exact one.
                    _this.previousPosition = {
                        length: x * x + y * y
                    };
                }
                _this.onMousedown(event);
            };
            this.onMouseMove = function (event) {
                // console.log('PanZoomComponent: onMouseMove()', event);
                // console.log(`PanZoomComponent: onMouseMove(): event.timeStamp:`, event.timeStamp);
                // timestamp - 10587.879999999132 - milliseconds
                // Called when moving the mouse with the left button down
                var evWrap;
                if (event && event.pageX && event.pageY) {
                    evWrap = event;
                }
                else if ((event.type === 'touchmove') && event.touches[0].pageX && event.touches[0].pageY) {
                    evWrap = {
                        pageX: event.touches[0].pageX,
                        pageY: event.touches[0].pageY
                    };
                }
                else {
                    return;
                }
                var now = event.timeStamp;
                var timeSinceLastMouseEvent = (now - _this.lastMouseEventTime) / 1000;
                _this.lastMouseEventTime = now;
                var dragDelta = {
                    // a representation of how far each coordinate has moved since the last time it was moved
                    x: evWrap.pageX - _this.previousPosition.x,
                    y: evWrap.pageY - _this.previousPosition.y
                };
                if (_this.config.keepInBounds) {
                    var topLeftCornerView = _this.getViewPosition({ x: 0, y: 0 });
                    var bottomRightCornerView = _this.getViewPosition({ x: _this.contentWidth, y: _this.contentHeight });
                    if (topLeftCornerView.x > 0 && dragDelta.x > 0) {
                        dragDelta.x *= Math.min(1, Math.pow(topLeftCornerView.x, -_this.config.keepInBoundsDragPullback));
                    }
                    if (topLeftCornerView.y > 0 && dragDelta.y > 0) {
                        dragDelta.y *= Math.min(1, Math.pow(topLeftCornerView.y, -_this.config.keepInBoundsDragPullback));
                    }
                    if (bottomRightCornerView.x < _this.contentWidth && dragDelta.x < 0) {
                        dragDelta.x *= Math.min(1, Math.pow(_this.contentWidth - bottomRightCornerView.x, -_this.config.keepInBoundsDragPullback));
                    }
                    if (bottomRightCornerView.y < _this.contentHeight && dragDelta.y < 0) {
                        dragDelta.y *= Math.min(1, Math.pow(_this.contentHeight - bottomRightCornerView.y, -_this.config.keepInBoundsDragPullback));
                    }
                }
                // now pan the view
                var delta = {
                    x: dragDelta.x || 0,
                    y: dragDelta.y || 0
                };
                _this.model.pan.x += delta.x;
                _this.model.pan.y += delta.y;
                _this.syncBaseToModel();
                _this.animationTick();
                if (!_this.model.isPanning) {
                    // This will improve the performance,
                    // because the browser stops evaluating hits against the elements displayed inside the pan zoom view.
                    // Besides this, mouse events will not be sent to any other elements,
                    // this prevents issues like selecting elements while dragging.
                    _this.panzoomOverlayRef.nativeElement.style.display = 'block';
                }
                _this.model.isPanning = true;
                // set these for the animation slow down once drag stops
                // panVelocity is a measurement of speed for x and y coordinates, in pixels per mouse move event.  It is a measure of how fast the mouse is moving
                var panVelocityX = dragDelta.x / timeSinceLastMouseEvent;
                var panVelocityY = dragDelta.y / timeSinceLastMouseEvent;
                _this.panVelocity = {
                    x: !Number.isFinite(panVelocityX) ? 0 : panVelocityX,
                    y: !Number.isFinite(panVelocityY) ? 0 : panVelocityY
                };
                // console.log(`PanZoomComponent: onMouseMove(): panVelocity:`, this.panVelocity);
                _this.previousPosition = {
                    x: evWrap.pageX,
                    y: evWrap.pageY
                };
            };
            this.onTouchMove = function (event) {
                // console.log('PanZoomComponent: onTouchMove()');
                // console.log('PanZoomComponent: onTouchMove(): event:', event);
                if (event.touches.length === 1) {
                    // single touch, emulate mouse move
                    _this.onMouseMove(event);
                }
                else {
                    // multiple touches, zoom in/out
                    // console.log('pinch zooming');
                    // Calculate x and y distance between touch events
                    var x = event.touches[0].pageX - event.touches[1].pageX;
                    var y = event.touches[0].pageY - event.touches[1].pageY;
                    // Calculate length between touch points with pythagoras
                    // There is no reason to use Math.pow and Math.sqrt as we
                    // only want a relative length and not the exact one.
                    var length = x * x + y * y;
                    // Calculate delta between current position and last position
                    var delta = length - _this.previousPosition.length;
                    // Naive hysteresis
                    if (Math.abs(delta) < 100) {
                        return;
                    }
                    // Calculate center between touch points
                    var centerX = event.touches[1].pageX + x / 2;
                    var centerY = event.touches[1].pageY + y / 2;
                    // Calculate zoom center
                    var frameElementOffset = _this.getFrameElementOffset();
                    var clickPoint = {
                        x: centerX - frameElementOffset.left,
                        y: centerY - frameElementOffset.top
                    };
                    _this.lastClickPoint = clickPoint;
                    _this.changeZoomLevel(_this.base.zoomLevel + delta * 0.0001, clickPoint);
                    // Update length for next move event
                    _this.previousPosition = {
                        length: length
                    };
                }
            };
            this.onMouseUp = function (event) {
                // console.log('PanZoomComponent: onMouseup()', event);
                if ((event.button !== _this.dragMouseButton) && (event.type !== 'touchend')) {
                    return;
                }
                if (event.type !== 'touchend') {
                    event.preventDefault();
                }
                var now = event.timeStamp;
                var timeSinceLastMouseEvent = (now - _this.lastMouseEventTime) / 1000;
                if (_this.panVelocity && (_this.panVelocity.x !== 0 || _this.panVelocity.y !== 0)) {
                    // apply strong initial dampening if the mouse up occured much later than the last mouse move, indicating that the mouse hasn't moved recently
                    // TBD - experiment with this formula
                    var initialMultiplier = Math.max(0, -0.2 + Math.pow(timeSinceLastMouseEvent + 1, -4));
                    _this.panVelocity.x *= initialMultiplier;
                    _this.panVelocity.y *= initialMultiplier;
                    _this.dragFinishing = true;
                    _this.zone.runOutsideAngular(function () { return _this.animationId = _this.animationFrameFunc(_this.animationTick); });
                }
                else {
                    _this.panVelocity = undefined;
                    _this.dragFinishing = false;
                    _this.model.isPanning = false;
                    _this.config.modelChanged.next(_this.model);
                    _this.syncBaseToModel();
                }
                _this.isDragging = false;
                if (_this.isMobile) {
                    _this.zone.runOutsideAngular(function () { return document.removeEventListener('touchend', _this.onTouchEnd); });
                    _this.zone.runOutsideAngular(function () { return document.removeEventListener('touchmove', _this.onTouchMove, { passive: true, capture: false }); });
                }
                else {
                    _this.zone.runOutsideAngular(function () { return document.removeEventListener('mousemove', _this.onMouseMove, { passive: true, capture: false }); });
                    _this.zone.runOutsideAngular(function () { return document.removeEventListener('mouseup', _this.onMouseUp, { passive: true }); });
                }
                // Set the overlay to non-blocking again:
                _this.panzoomOverlayRef.nativeElement.style.display = 'none';
            };
            this.onTouchEnd = function (event) {
                // console.log('PanZoomComponent: onTouchEnd()');
                _this.onMouseUp(event);
            };
            this.onDblClick = function (event) {
                // console.log('PanZoomComponent: onDblClick()');
                event.preventDefault();
                if (!_this.config.zoomOnDoubleClick) {
                    return;
                }
                var frameElementOffset = _this.getFrameElementOffset();
                var clickPoint = {
                    x: event.pageX - frameElementOffset.left,
                    y: event.pageY - frameElementOffset.top
                };
                _this.lastClickPoint = clickPoint;
                _this.zoomInToPoint(clickPoint);
            };
            this.preventDefault = function (event) {
                event.preventDefault();
            };
            ////////////////////////////////////////////////////////////////////////////////
            ///////////////////////////// END EVENT HANDLERS ///////////////////////////////
            ////////////////////////////////////////////////////////////////////////////////
            ////////////////////////////////////////////////////////////////////////////////
            ///////////////////////////// APPLY ANIMATIONS /////////////////////////////////
            ////////////////////////////////////////////////////////////////////////////////
            this.animationTick = function (timestamp) {
                if (timestamp === void 0) { timestamp = performance.now(); }
                // console.log('PanZoomComponent: animationTick()');
                // timestamp looks like 76916.963.  The unit is milliseconds and should be accurate to 5 µs (microseconds)
                var deltaTime = 0;
                if (_this.lastTick !== 0) {
                    deltaTime = (timestamp - _this.lastTick); // orig - milliseconds since the last animationTick
                }
                _this.lastTick = timestamp;
                if (_this.animationParams) {
                    // when we're running an animation (but not waiting for a released drag to halt)
                    // console.log('PanZoomComponent: animationTick(): model is zooming');
                    _this.animationParams.progress += Math.abs(deltaTime / _this.animationParams.duration);
                    if (_this.animationParams.progress >= 1.0) {
                        // Only when the animation has finished, sync the base to the model.
                        _this.animationParams.progress = 1.0;
                        _this.updateDOM();
                        _this.animationParams = undefined;
                    }
                }
                if (_this.panVelocity && _this.dragFinishing) {
                    // This is when we've panned and released the mouse button and the view is "free-floating" until it slows to a halt.  Don't let the while loop fool you - this only applies it for the current frame.
                    // Prevent overshooting if delta time is large for some reason. We apply the simple solution of slicing delta time into smaller pieces and applying each one
                    if (deltaTime > 0) {
                        deltaTime = deltaTime / 1000;
                    }
                    while (deltaTime > 0) {
                        var dTime = Math.min(.02, deltaTime);
                        deltaTime = deltaTime - dTime;
                        _this.model.pan.x = _this.model.pan.x + _this.panVelocity.x * dTime;
                        _this.panVelocity.x = _this.panVelocity.x * (1 - _this.config.friction * dTime);
                        _this.model.pan.y = _this.model.pan.y + _this.panVelocity.y * dTime;
                        _this.panVelocity.y = _this.panVelocity.y * (1 - _this.config.friction * dTime);
                        var speed = _this.length(_this.panVelocity);
                        if (speed <= _this.config.haltSpeed) {
                            _this.panVelocity = undefined;
                            _this.dragFinishing = false;
                            break;
                        }
                    }
                }
                if (_this.config.keepInBounds || _this.dragFinishing) {
                    // Checks that keepInBounds is set and that the mouse button isn't pressed, and if so, it stops the contents from going out of view
                    // console.log('PanZoomComponent: animationTick(): keepInBounds');
                    var topLeftCornerView = _this.getViewPosition({ x: 0, y: 0 });
                    var bottomRightCornerView = _this.getViewPosition({ x: _this.contentWidth, y: _this.contentHeight });
                    if (topLeftCornerView.x > 0) {
                        _this.base.pan.x -= _this.config.keepInBoundsRestoreForce * topLeftCornerView.x;
                    }
                    if (topLeftCornerView.y > 0) {
                        _this.base.pan.y -= _this.config.keepInBoundsRestoreForce * topLeftCornerView.y;
                    }
                    if (bottomRightCornerView.x < _this.contentWidth) {
                        _this.base.pan.x -= _this.config.keepInBoundsRestoreForce * (bottomRightCornerView.x - _this.contentWidth);
                    }
                    if (bottomRightCornerView.y < _this.contentHeight) {
                        _this.base.pan.y -= _this.config.keepInBoundsRestoreForce * (bottomRightCornerView.y - _this.contentHeight);
                    }
                }
                _this.updateDOM();
                _this.config.modelChanged.next(_this.model);
                if (_this.animationParams || (_this.panVelocity && _this.dragFinishing)) {
                    // Is an animation active?  If so, run the next frame
                    if (_this.isChrome && _this.zoomLevelIsChanging) {
                        // run will-change toggle hack on Chrome to trigger re-rasterization
                        // see https://developers.google.com/web/updates/2016/09/re-rastering-composite
                        if (_this.willChangeNextFrame) {
                            _this.zoomElementRef.nativeElement.style.willChange = 'auto';
                        }
                        else {
                            _this.zoomElementRef.nativeElement.style.willChange = 'transform';
                        }
                        _this.willChangeNextFrame = !_this.willChangeNextFrame;
                    }
                    _this.animationFrameFunc(_this.animationTick); // Call the next animation frame
                }
                else if (_this.panVelocity && !_this.dragFinishing) {
                    // we're just mouse-panning the frame.  We don't need another tick
                    return;
                }
                else {
                    // Animation has ended
                    if (_this.model.isPanning) {
                        _this.model.isPanning = false;
                    }
                    _this.syncBaseToModel();
                    _this.config.modelChanged.next(_this.model);
                    _this.scale = _this.getCssScale(_this.base.zoomLevel);
                    _this.willChangeNextFrame = true;
                    _this.zoomElementRef.nativeElement.style.willChange = 'transform';
                    _this.zoomLevelIsChanging = false;
                    _this.lastTick = 0;
                }
            };
        }
        PanZoomComponent.prototype.ngOnInit = function () {
            var _this = this;
            // console.log('PanZoomComponent: ngOnInit(): initializing PanZoomComponent');
            if (this.config.initialZoomToFit) {
                this.base = this.calcZoomToFit(this.config.initialZoomToFit);
            }
            else {
                this.base = {
                    zoomLevel: this.config.initialZoomLevel,
                    pan: {
                        x: this.config.initialPanX,
                        y: this.config.initialPanY
                    }
                };
            }
            this.model = {
                zoomLevel: this.base.zoomLevel,
                isPanning: false,
                pan: {
                    x: this.base.pan.x,
                    y: this.base.pan.y
                }
            };
            this.config.modelChanged.next(this.model);
            // create public API
            this.api = {
                model: this.model,
                config: this.config,
                changeZoomLevel: this.zoomToLevelAndPoint.bind(this),
                zoomIn: this.zoomIn.bind(this),
                zoomOut: this.zoomOut.bind(this),
                zoomToFit: this.zoomToFit.bind(this),
                resetView: this.resetView.bind(this),
                getViewPosition: this.getViewPosition.bind(this),
                getModelPosition: this.getModelPosition.bind(this),
                panToPoint: this.panToPoint.bind(this),
                panDelta: this.panDelta.bind(this),
                panDeltaPercent: this.panDeltaPercent.bind(this),
                panDeltaAbsolute: this.panDeltaAbsolute.bind(this),
                centerContent: this.centerContent.bind(this),
                centerX: this.centerX.bind(this),
                centerY: this.centerY.bind(this),
                centerTopLeft: this.centerTopLeftCorner.bind(this),
                centerBottomLeft: this.centerBottomLeftCorner.bind(this),
                centerTopRight: this.centerTopRightCorner.bind(this),
                centerBottomRight: this.centerBottomRightCorner.bind(this),
                updateContentDimensions: this.updateContentDimensions.bind(this),
                detectContentDimensions: this.detectContentDimensions.bind(this)
            };
            this.config.api.next(this.api);
            if (this.config.freeMouseWheel) {
                this.scale = this.getCssScale(this.config.initialZoomLevel);
                var maxZoomLevel = this.config.zoomLevels - 1;
                this.maxScale = this.getCssScale(maxZoomLevel);
                this.minScale = this.getCssScale(0);
            }
            this.minimumAllowedZoomLevel = 0;
            if (this.config.keepInBounds) {
                this.minimumAllowedZoomLevel = this.config.neutralZoomLevel;
                this.minScale = this.getCssScale(this.config.neutralZoomLevel);
            }
            this.zoomElementRef.nativeElement.style.willChange = 'transform';
            if (navigator.userAgent.search('Chrome') >= 0) {
                this.isChrome = true;
                this.zoomElementRef.nativeElement.style.transform = 'translateZ(0)';
            }
            if (this.config.acceleratePan) {
                this.panElementRef.nativeElement.style.willChange = 'transform';
                if (navigator.userAgent.search('Chrome') >= 0) {
                    this.isChrome = true;
                    this.panElementRef.nativeElement.style.transform = 'translateZ(0)';
                }
            }
            this.animationTick();
            this.scale = this.getCssScale(this.base.zoomLevel);
            this.isFirstSync = false;
            switch (this.config.dragMouseButton) {
                case 'left':
                    this.dragMouseButton = 0;
                    break;
                case 'middle':
                    this.dragMouseButton = 1;
                    this.zone.runOutsideAngular(function () { return _this.frameElementRef.nativeElement.addEventListener('auxclick', _this.preventDefault); });
                    break;
                case 'right':
                    this.zone.runOutsideAngular(function () { return document.addEventListener('contextmenu', _this.preventDefault); });
                    this.dragMouseButton = 2;
                    break;
                default:
                    this.dragMouseButton = 0; // left
            }
        };
        PanZoomComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            // console.log('PanZoomComponent: ngAfterViewInit()');
            this.detectContentDimensions();
            var frameStyle = getComputedStyle(this.frameElementRef.nativeElement);
            this.frameHeight = parseInt(frameStyle.getPropertyValue('height').split('px')[0]);
            this.frameWidth = parseInt(frameStyle.getPropertyValue('width').split('px')[0]);
            this.zone.runOutsideAngular(function () { return _this.animationFrameFunc = window.requestAnimationFrame; });
            if (this.config.dynamicContentDimensions) {
                if (window.ResizeObserver) {
                    this.resizeObserver = new window.ResizeObserver(function (entries) { return _this.onContentDimensionsChangeDetected(entries); });
                    this.zone.runOutsideAngular(function () { return _this.resizeObserver.observe(_this.zoomElementRef.nativeElement); });
                }
                else {
                    console.error('ResizeObserver API is not supported by this browser.  See https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver for info on browser compatibility.');
                }
            }
            if (this.isMobileDevice()) {
                this.isMobile = true;
                this.zone.runOutsideAngular(function () { return _this.frameElementRef.nativeElement.addEventListener('touchstart', _this.onTouchStart); });
            }
            else {
                this.zone.runOutsideAngular(function () {
                    _this.frameElementRef.nativeElement.addEventListener('mousedown', _this.onMousedown);
                    _this.frameElementRef.nativeElement.addEventListener('dblclick', _this.onDblClick);
                    _this.frameElementRef.nativeElement.addEventListener('wheel', function (event) { return _this.animationFrameFunc(function () { return _this.onMouseWheel(event); }); }, { passive: true });
                });
            }
        };
        PanZoomComponent.prototype.ngOnDestroy = function () {
            var _this = this;
            // console.log('PanZoomComponent: ngOnDestroy()');
            if (this.isMobile) {
                this.frameElementRef.nativeElement.removeEventListener('touchstart', this.onTouchStart);
            }
            else {
                this.frameElementRef.nativeElement.removeEventListener('mousedown', this.onMousedown);
                this.frameElementRef.nativeElement.removeEventListener('wheel', function (event) { return _this.animationFrameFunc(function () { return _this.onMouseWheel(event); }); }, { passive: true });
                this.frameElementRef.nativeElement.removeEventListener('dblclick', this.onDblClick);
            }
            if (this.animationFrameFunc && this.animationId) {
                window.cancelAnimationFrame(this.animationId);
            }
            if (this.resizeObserver) {
                this.resizeObserver.disconnect();
            }
            switch (this.config.dragMouseButton) {
                case 'middle':
                    this.dragMouseButton = 1;
                    this.zone.runOutsideAngular(function () { return _this.frameElementRef.nativeElement.removeEventListener('auxclick', _this.preventDefault); });
                    break;
                case 'right':
                    this.zone.runOutsideAngular(function () { return document.removeEventListener('contextmenu', _this.preventDefault); });
                    this.dragMouseButton = 2;
                    break;
            }
        };
        /**
         * Method will check if 'child' node element is a child of
         * parent node with class 'className'
         * @param className
         * @param child
         */
        PanZoomComponent.prototype.isParentElement = function (className, child) {
            var node = child.parentNode;
            while (node !== null) {
                if (node.classList && node.classList.contains(className)) {
                    return true;
                }
                node = node.parentNode;
            }
            return false;
        };
        PanZoomComponent.prototype.updateDOM = function () {
            // console.log('PanZoomComponent: updateDOM()');
            // Called by ngAfterViewInit() and animationTick()
            // This function does not get called by freeZoom(), which operates independently of animationTick() and updateDOM().
            if (this.animationParams) {
                // we're running an animation sequence (but not freeZooming or panning with onMouseMove() )
                this.model.zoomLevel = this.base.zoomLevel + this.animationParams.deltaZoomLevel * this.animationParams.progress; // calculate how far we need to zoom in or out for the current animationTick
                var deltaTranslation = this.animationParams.panStepFunc(this.model.zoomLevel); // calculate how far to pan the view to based on our translated coordinates
                // sync the model pan coordinates to our translated pan coordinates
                // we do this by adding how far we want to move in each direction to our our existing base pan coordinates (where we started)
                this.model.pan.x = this.base.pan.x + deltaTranslation.x;
                this.model.pan.y = this.base.pan.y + deltaTranslation.y;
                if (this.config.keepInBounds) {
                    var topLeftCornerView = this.getViewPosition({ x: 0, y: 0 });
                    var bottomRightCornerView = this.getViewPosition({ x: this.contentWidth, y: this.contentHeight });
                    if (topLeftCornerView.x > 0) {
                        this.model.pan.x = 0;
                    }
                    if (topLeftCornerView.y > 0) {
                        this.model.pan.y = 0;
                    }
                    if (bottomRightCornerView.x < this.contentWidth) {
                        this.model.pan.x -= (bottomRightCornerView.x - this.contentWidth);
                    }
                    if (bottomRightCornerView.y < this.contentHeight) {
                        this.model.pan.y -= (bottomRightCornerView.y - this.contentHeight);
                    }
                }
            }
            ////////////////////////////////////////////////////
            //////////////////// APPLY SCALING /////////////////
            ////////////////////////////////////////////////////
            if (this.animationParams || this.isFirstSync) {
                var scale = this.getCssScale(this.model.zoomLevel);
                var scaleString = "scale(" + scale + ")";
                var zoomStyle = "transform-origin: 0 0; transform: " + scaleString + ";";
                this.zoomElementRef.nativeElement.setAttribute('style', zoomStyle);
            }
            ////////////////////////////////////////////////////
            //////////////// APPLY PAN ANIMATION ///////////////
            ////////////////////////////////////////////////////
            var translate3d = "translate3d(" + this.model.pan.x + "px, " + this.model.pan.y + "px, 0)";
            this.panElementRef.nativeElement.style.transform = translate3d;
        };
        PanZoomComponent.prototype.freeZoom = function (clickPoint, wheelDelta) {
            // console.log('PanZoomComponent: freeZoom(): this.base:', this.base);
            if (this.isDragging) {
                // don't allow zooming if the mouse is down
                return;
            }
            // now handle interruption of an in-progress animation
            if (this.animationParams) {
                this.animationParams = undefined; // cancel an existing animation
            }
            if (this.panVelocity) {
                this.dragFinishing = false;
                this.panVelocity = undefined;
            }
            var currentPan = {
                // the current model coordinates
                x: this.model.pan.x,
                y: this.model.pan.y
            };
            var currentScale = this.scale; // get the current CSS scale (scale0)
            var newScale = this.scale + (wheelDelta * this.config.freeMouseWheelFactor * this.scale);
            // takes either the minimum scale (furthest allowable zoomed out) or the calculated current scale, whichever is greater, unless calculated current scale exceeds maxScale (furthest allowable zoomed in), in which case maxScale is used
            newScale = Math.max(this.minScale, Math.min(this.maxScale, newScale));
            this.scale = newScale;
            var targetPoint = {
                // The target point to zoom to.  It must stay the same as the untranslated point
                x: clickPoint.x - (newScale / currentScale) * (clickPoint.x - currentPan.x),
                y: clickPoint.y - (newScale / currentScale) * (clickPoint.y - currentPan.y)
            };
            // Apply Pan & Scale
            var translate3d = "translate3d(" + targetPoint.x + "px, " + targetPoint.y + "px, 0)";
            this.panElementRef.nativeElement.style.transform = translate3d;
            var scaleString = "scale(" + this.scale + ")";
            var zoomStyle = "transform-origin: 0 0; transform: " + scaleString + ";";
            if (this.isChrome) {
                if (this.willChangeNextFrame) {
                    zoomStyle += " will-change: auto;";
                }
                else {
                    zoomStyle += " will-change: transform;";
                }
                this.willChangeNextFrame = !this.willChangeNextFrame;
            }
            // apply zoom css
            this.zoomElementRef.nativeElement.setAttribute('style', zoomStyle);
            this.model.pan.x = targetPoint.x;
            this.model.pan.y = targetPoint.y;
            this.model.zoomLevel = this.getZoomLevel(this.scale);
            this.config.modelChanged.next(this.model);
            this.syncBaseToModel();
        };
        ////////////////////////////////////////////////////
        //////////////// HELPER FUNCTIONS //////////////////
        ////////////////////////////////////////////////////
        PanZoomComponent.prototype.isMobileDevice = function () {
            return (typeof window.orientation !== 'undefined') || (navigator.userAgent.indexOf('IEMobile') !== -1);
        };
        PanZoomComponent.prototype.syncBaseToModel = function () {
            this.base.pan.x = this.model.pan.x;
            this.base.pan.y = this.model.pan.y;
            this.base.zoomLevel = this.model.zoomLevel;
        };
        PanZoomComponent.prototype.length = function (vector2d) {
            // console.log('PanZoomComponent: length()');
            return Math.sqrt(vector2d.x * vector2d.x + vector2d.y * vector2d.y);
        };
        PanZoomComponent.prototype.getCentrePoint = function () {
            // console.log('PanZoomComponent: getCentrePoint()');
            return {
                x: this.frameElementRef.nativeElement.offsetWidth / 2,
                y: this.frameElementRef.nativeElement.offsetHeight / 2
            };
        };
        PanZoomComponent.prototype.getCssScale = function (zoomLevel) {
            // console.log('PanZoomComponent: getCssScale()');
            return Math.pow(this.config.scalePerZoomLevel, zoomLevel - this.config.neutralZoomLevel);
        };
        PanZoomComponent.prototype.getZoomLevel = function (cssScale) {
            // console.log('PanZoomComponent: getZoomLevel()');
            return Math.log10(cssScale) / Math.log10(this.config.scalePerZoomLevel) + this.config.neutralZoomLevel;
        };
        PanZoomComponent.prototype.calcZoomToFit = function (rect) {
            // console.log('PanZoomComponent: calcZoomToFit(): rect:', rect);
            // let (W, H) denote the size of the viewport
            // let (w, h) denote the size of the rectangle to zoom to
            // then we must CSS scale by the min of W/w and H/h in order to just fit the rectangle
            // returns the target left and top coordinates for the panElement and target zoomLevel
            var viewportWidth = this.frameElementRef.nativeElement.offsetWidth;
            var viewportHeight = this.frameElementRef.nativeElement.offsetHeight;
            var targetWidth = rect.width;
            var targetHeight = rect.height;
            var cssScaleExact = Math.min(viewportWidth / targetWidth, viewportHeight / targetHeight);
            var zoomLevelExact = this.getZoomLevel(cssScaleExact);
            var zoomLevel = zoomLevelExact * this.config.zoomToFitZoomLevelFactor;
            var cssScale = this.getCssScale(zoomLevel);
            return {
                zoomLevel: zoomLevel,
                pan: {
                    x: -rect.x * cssScale + (viewportWidth - targetWidth * cssScale) / 2,
                    y: -rect.y * cssScale + (viewportHeight - targetHeight * cssScale) / 2
                }
            };
        };
        PanZoomComponent.prototype.zoomToFitModel = function (target, duration) {
            if (duration === void 0) { duration = undefined; }
            // console.log('PanZoomComponent: zoomToFitModel(): target:', target);
            // target.pan.x is the panElement left style property
            // target.pan.y is the panElement top style property
            this.model.isPanning = true;
            this.animateToTarget(target, duration);
        };
        PanZoomComponent.prototype.zoomToLevelAndPoint = function (level, clickPoint) {
            // console.log('PanZoomComponent: zoomToLevelAndPoint(): level:', level);
            // console.log('PanZoomComponent: zoomToLevelAndPoint(): clickPoint:', clickPoint);
            this.changeZoomLevel(level, clickPoint);
        };
        PanZoomComponent.prototype.zoomIn = function (zoomType) {
            if (zoomType === void 0) { zoomType = 'lastPoint'; }
            // console.log('PanZoomComponent: zoomIn()');
            if (zoomType === 'lastPoint') {
                this.changeZoomLevel(this.base.zoomLevel + this.config.zoomButtonIncrement, this.lastClickPoint);
            }
            else if (zoomType === 'viewCenter') {
                this.changeZoomLevel(this.base.zoomLevel + this.config.zoomButtonIncrement, this.getCentrePoint());
            }
        };
        PanZoomComponent.prototype.zoomOut = function (zoomType) {
            if (zoomType === void 0) { zoomType = 'lastPoint'; }
            // console.log('PanZoomComponent: zoomOut()');
            if (zoomType === 'lastPoint') {
                this.changeZoomLevel(this.base.zoomLevel - this.config.zoomButtonIncrement, this.lastClickPoint);
            }
            else if (zoomType === 'viewCenter') {
                this.changeZoomLevel(this.base.zoomLevel - this.config.zoomButtonIncrement, this.getCentrePoint());
            }
        };
        PanZoomComponent.prototype.startAnimation = function () {
            var _this = this;
            this.lastTick = performance.now();
            this.zone.runOutsideAngular(function () { return _this.animationId = _this.animationFrameFunc(_this.animationTick); });
        };
        PanZoomComponent.prototype.getFrameElementOffset = function () {
            var rect = this.frameElementRef.nativeElement.getBoundingClientRect();
            return {
                top: rect.top + window.scrollY,
                left: rect.left + window.scrollX
            };
        };
        PanZoomComponent.prototype.onContentDimensionsChangeDetected = function (entries) {
            // callback for ResizeObserver
            this.contentHeight = entries[0].contentRect.height;
            this.contentWidth = entries[0].contentRect.width;
        };
        ////////////////////////////////////////////////////
        /////////////////// API METHODS ////////////////////
        ////////////////////////////////////////////////////
        PanZoomComponent.prototype.getViewPosition = function (modelPosition) {
            // console.log('PanZoomComponent: getViewPosition()');
            // p' = p * s + t
            // viewPosition = modelPosition * scale + basePan
            var scale, translation;
            if (this.animationParams) {
                scale = this.getCssScale(this.base.zoomLevel + this.animationParams.deltaZoomLevel * this.animationParams.progress);
                var deltaTranslation = this.animationParams.panStepFunc(this.model.zoomLevel);
                translation = {
                    x: this.base.pan.x + deltaTranslation.x,
                    y: this.base.pan.y + deltaTranslation.y
                };
            }
            else {
                scale = this.getCssScale(this.base.zoomLevel);
                translation = this.base.pan;
            }
            return {
                x: modelPosition.x * scale + translation.x,
                y: modelPosition.y * scale + translation.y
            };
        };
        PanZoomComponent.prototype.getModelPosition = function (viewPosition) {
            // console.log('PanZoomComponent: getModelPosition()');
            // p = (1/s)(p' - t)
            var pmark = viewPosition;
            var s = this.getCssScale(this.base.zoomLevel);
            var t = this.base.pan;
            return {
                x: (1 / s) * (pmark.x - t.x),
                y: (1 / s) * (pmark.y - t.y)
            };
        };
        PanZoomComponent.prototype.resetView = function () {
            // console.log('PanZoomComponent: resetView()');
            if (this.config.initialZoomToFit) {
                this.zoomToFit(this.config.initialZoomToFit);
            }
            else if (this.config.initialPanX !== undefined && this.config.initialPanY !== undefined && this.config.initialZoomLevel !== undefined) {
                this.zoomToFitModel({
                    zoomLevel: this.config.initialZoomLevel,
                    pan: {
                        x: this.config.initialPanX,
                        y: this.config.initialPanY
                    }
                });
            }
            else {
                console.error('PanZoomComponent: resetView() could not reset view as some vars were not set.  The culprits are either config.initialZoomLevel, config.initialPanX, or config.initialPanY.  Or just set panzoomConfig.initialZoomToFit');
                console.log('config.initialZoomLevel: ' + this.config.initialZoomLevel);
                console.log('config.initialPanX: ' + this.config.initialPanX);
                console.log('config.initialPanY: ' + this.config.initialPanY);
            }
        };
        PanZoomComponent.prototype.zoomToFit = function (rectangle, duration) {
            // console.log('PanZoomComponent: zoomToFit(): rectangle', rectangle);
            // when a user clicks a zoom to fit button
            // example rectangle: { "x": 0, "y": 100, "width": 100, "height": 100 }
            var target = this.calcZoomToFit(rectangle);
            // target.pan.x is the panElement left style property
            // target.pan.y is the panElement top style property
            this.model.isPanning = true;
            this.animateToTarget(target, duration);
        };
        PanZoomComponent.prototype.zoomInToPoint = function (clickPoint) {
            // console.log('PanZoomComponent: zoomIn(): clickPoint:', clickPoint);
            this.changeZoomLevel(this.base.zoomLevel + this.config.zoomButtonIncrement, clickPoint);
        };
        PanZoomComponent.prototype.zoomOutFromPoint = function (clickPoint) {
            // console.log('PanZoomComponent: zoomOut()');
            this.changeZoomLevel(this.base.zoomLevel - this.config.zoomButtonIncrement, clickPoint);
        };
        PanZoomComponent.prototype.panToPoint = function (point, duration) {
            // console.log('PanZoomComponent: panToPoint(): point:', point);
            // API call to animate the view so that the centre point of the view is at the
            // point parameter coordinates, relative to the original, unzoomed
            // content width and height
            // example point: { "x": 0, "y": 0 } // makes the top-left corner of the content
            // the centre of the view
            var target = {
                pan: {
                    x: (this.frameWidth / 2) - point.x * this.scale,
                    y: (this.frameHeight / 2) - point.y * this.scale
                },
                zoomLevel: this.base.zoomLevel
            };
            this.model.isPanning = true;
            this.animateToTarget(target, duration);
        };
        PanZoomComponent.prototype.panToPointCurrentScale = function (point, duration) {
            // console.log('PanZoomComponent: panToPoint(): point:', point);
            // API call to animate the view so that the centre point of the view is at the
            // point parameter coordinates, relative to the original, unzoomed
            // content width and height
            // example point: { "x": 0, "y": 0 } // makes the top-left corner of the content
            // the centre of the view
            var target = {
                pan: {
                    x: (this.frameWidth / 2) - point.x,
                    y: (this.frameHeight / 2) - point.y
                },
                zoomLevel: this.base.zoomLevel
            };
            this.model.isPanning = true;
            this.animateToTarget(target, duration);
        };
        PanZoomComponent.prototype.panDelta = function (delta, duration) {
            // console.log('PanZoomComponent: panDelta(): delta:', delta);
            // API call to pan the view left, right, up, or down, based on a number of pixels
            // of the original, unzoomed content.
            // Positive is right and down
            // Negative is left and up
            // example point: { "x": 100, "y": -50 } // moves the view right 50px and up 50px
            var target = {
                pan: {
                    x: this.base.pan.x - this.scale * delta.x,
                    y: this.base.pan.y - this.scale * delta.y
                },
                zoomLevel: this.base.zoomLevel
            };
            this.model.isPanning = true;
            this.animateToTarget(target, duration);
        };
        PanZoomComponent.prototype.panDeltaAbsolute = function (delta, duration) {
            // console.log('PanZoomComponent: panDeltaAbsolute(): delta:', delta);
            // API call to pan the view left, right, up, or down, based on a number of pixels
            // This method doesn't adjust for scale.  I'm not sure why you'd want this
            // but have it here just in case someone needs it
            // Positive is right and down
            // Negative is left and up
            // example point: { "x": 100, "y": -50 } // moves the view right 50px and up 50px
            var target = {
                pan: {
                    x: this.base.pan.x - delta.x,
                    y: this.base.pan.y - delta.y
                },
                zoomLevel: this.base.zoomLevel
            };
            this.model.isPanning = true;
            this.animateToTarget(target, duration);
        };
        PanZoomComponent.prototype.panDeltaPercent = function (deltaPercent, duration) {
            // console.log('PanZoomComponent: panDeltaPercent(): deltaPercent:', deltaPercent);
            // API call to pan the view up, down, left, or right, based on a percentage
            // of the original, unzoomed content width and height
            // example point: { "x": 10, "y": -20 }
            var deltaX = 0;
            var deltaY = 0;
            if (deltaPercent.x !== 0) {
                deltaX = this.contentWidth * (deltaPercent.x / 100) * this.scale;
            }
            if (deltaPercent.y !== 0) {
                deltaY = this.contentHeight * (deltaPercent.y / 100) * this.scale;
            }
            var target = {
                pan: {
                    x: this.base.pan.x - deltaX,
                    y: this.base.pan.y - deltaY
                },
                zoomLevel: this.base.zoomLevel
            };
            // target.pan.x is the panElement left style property
            // target.pan.y is the panElement top style property
            this.model.isPanning = true;
            this.animateToTarget(target, duration);
        };
        PanZoomComponent.prototype.centerContent = function (duration) {
            this.panToPoint({
                x: this.contentWidth / 2,
                y: this.contentHeight / 2
            }, duration);
        };
        PanZoomComponent.prototype.centerX = function (duration) {
            var target = {
                pan: {
                    x: (this.frameWidth / 2) - (this.contentWidth / 2) * this.scale,
                    y: this.base.pan.y
                },
                zoomLevel: this.base.zoomLevel
            };
            this.model.isPanning = true;
            this.animateToTarget(target, duration);
        };
        PanZoomComponent.prototype.centerY = function (duration) {
            var target = {
                pan: {
                    x: this.base.pan.x,
                    y: (this.frameHeight / 2) - (this.contentHeight / 2) * this.scale
                },
                zoomLevel: this.base.zoomLevel
            };
            this.model.isPanning = true;
            this.animateToTarget(target, duration);
        };
        PanZoomComponent.prototype.centerTopLeftCorner = function (duration) {
            this.panToPoint({
                x: 0,
                y: 0
            }, duration);
        };
        PanZoomComponent.prototype.centerBottomLeftCorner = function (duration) {
            this.panToPoint({
                x: 0,
                y: this.contentHeight
            }, duration);
        };
        PanZoomComponent.prototype.centerTopRightCorner = function (duration) {
            this.panToPoint({
                x: this.contentWidth,
                y: 0
            }, duration);
        };
        PanZoomComponent.prototype.centerBottomRightCorner = function (duration) {
            this.panToPoint({
                x: this.contentWidth,
                y: this.contentHeight
            }, duration);
        };
        PanZoomComponent.prototype.updateContentDimensions = function (width, height) {
            if (height !== undefined) {
                this.contentHeight = height;
            }
            if (width !== undefined) {
                this.contentWidth = width;
            }
        };
        PanZoomComponent.prototype.detectContentDimensions = function () {
            var zoomFrameStyle = getComputedStyle(this.zoomElementRef.nativeElement);
            this.contentHeight = parseInt(zoomFrameStyle.getPropertyValue('height').split('px')[0]);
            this.contentWidth = parseInt(zoomFrameStyle.getPropertyValue('width').split('px')[0]);
        };
        ////////////////////////////////////////////////////////////////////////////////
        ///////////////////////////// ANIMATION BUILDERS ///////////////////////////////
        ////////////////////////////////////////////////////////////////////////////////
        PanZoomComponent.prototype.animateToTarget = function (targetModel, duration) {
            var _this = this;
            if (duration === void 0) { duration = undefined; }
            // console.log('PanZoomComponent: animateToTarget()');
            // what this function really does is take a target model, and then sets
            // this.animationParams with the parameters for the whole animation,
            // namely the delta zoomLevel
            // it is the responsibility of the caller to kick off the animation with a call to animationFrameFunc()
            if (this.animationParams) {
                // make the user wait for existing animation to finish before clicking
                return;
            }
            this.zoomLevelIsChanging = false;
            if (this.base.zoomLevel !== targetModel.zoomLevel) {
                this.zoomLevelIsChanging = true;
            }
            var deltaZoomLevel = targetModel.zoomLevel - this.base.zoomLevel; // deltaZoomLevel is the number of zoom levels we are changing here
            var oldBase = {
                // the current base coordinates
                x: this.base.pan.x,
                y: this.base.pan.y
            };
            this.model.pan.x = this.base.pan.x;
            this.model.pan.y = this.base.pan.y;
            var panStepFunc = function (zoomLevel) {
                // this function gets called during every animation tick in updateDOM(), to calculate where to move the model pan coordinates to (i.e. the translation) for that tick, zoomLevel is ignored within animateToTarget()
                var targetPoint = {
                    // The target point to zoom to for the current animation frame.  It must stay the same as the untranslated point
                    x: (oldBase.x - targetModel.pan.x) * _this.animationParams.progress,
                    y: (oldBase.y - targetModel.pan.y) * _this.animationParams.progress
                };
                return {
                    x: -targetPoint.x,
                    y: -targetPoint.y
                };
            };
            // now set the parameters of our new animation
            duration = duration ? duration * 1000 : this.config.zoomStepDuration * 1000;
            this.animationParams = {
                deltaZoomLevel: deltaZoomLevel,
                panStepFunc: panStepFunc,
                duration: duration,
                progress: 0.0
            };
            this.startAnimation();
        };
        PanZoomComponent.prototype.changeZoomLevel = function (newZoomLevel, clickPoint) {
            var _this = this;
            // console.log('PanZoomComponent: changeZoomLevel()');
            if (this.animationParams) {
                // let's let any current animation just finish
                return;
            }
            this.zoomLevelIsChanging = true;
            // keep zoom level in bounds
            newZoomLevel = Math.max(this.minimumAllowedZoomLevel, newZoomLevel);
            newZoomLevel = Math.min(this.config.zoomLevels - 1, newZoomLevel);
            // console.log('newZoomLevel:', newZoomLevel);
            var deltaZoomLevel = newZoomLevel - this.base.zoomLevel; // deltaZoomLevel is the number of zoom levels we are changing here
            if (!deltaZoomLevel) {
                // a deltaZoomLevel of zero means that we aren't changing zoom, because we're either zoomed all the way in or all the way out
                return;
            }
            //
            // Let p be the vector to the clicked point in view coords and let p' be the same point in model coords. Let s be a scale factor
            // and let t be a translation vector. Let the transformation be defined as:
            //
            //  p' = p * s + t
            //
            // And conversely:
            //
            //  p = (1/s)(p' - t)
            //
            // Now use subscription 0 to denote the value before transform and zoom and let 1 denote the value after transform. Scale
            // changes from s0 to s1. Translation changes from t0 to t1. But keep p and p' fixed so that the view coordinate p' still
            // corresponds to the model coordinate p. This can be expressed as an equation relying upon solely upon p', s0, s1, t0, and t1:
            //
            //  (1/s0)(p - t0) = (1/s1)(p - t1)
            //
            // Every variable but t1 is known, thus it is easily isolated to:
            //
            //  t1 = p' - (s1/s0)*(p' - t0)
            //
            var currentPan = {
                // t0 - the current base coordinates
                x: this.base.pan.x,
                y: this.base.pan.y
            };
            var currentScale = this.scale; // s0 - get the current CSS scale (scale0)
            var destPoint = clickPoint || this.getCentrePoint(); // pmark - the point we are aiming to zoom to (either the click point or the centre of the page)
            var panStepFunc = function (zoomLevel) {
                // this function gets called during every animation tick, to calculate where to move the model pan coordinates to (i.e. the translation) for that tick, where zoomLevel is calculated from the current zoomLevel + the target zoomLevel * the progress of the current animation
                var targetScale = _this.getCssScale(zoomLevel); // s1 - the scale to translate to for the current animation tick
                var targetPoint = {
                    // t1 - The target point to pan to.  It must stay the same as the untranslated point
                    x: destPoint.x - (targetScale / currentScale) * (destPoint.x - currentPan.x),
                    y: destPoint.y - (targetScale / currentScale) * (destPoint.y - currentPan.y)
                };
                return {
                    // now return the difference between our initial click point and our translated (zoomed) click point
                    // these are not absolute coordinates - just how far to move them
                    x: targetPoint.x - currentPan.x,
                    y: targetPoint.y - currentPan.y
                };
            };
            // now set the parameters of our new animation
            this.animationParams = {
                deltaZoomLevel: deltaZoomLevel,
                panStepFunc: panStepFunc,
                duration: this.config.zoomStepDuration * 1000,
                progress: 0.0
            };
            this.startAnimation();
        };
        return PanZoomComponent;
    }());
    PanZoomComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'pan-zoom',
                    // we don't want to kill change detection for all elements beneath this, so we don't set OnPush.  Child views can implement OnPush if the developer wants to.  We can get away with this because the 'wheel' event handler runs outside of Angular, therefore it doesnt trigger change detection.
                    template: "<div\n  #frameElement\n  class=\"pan-zoom-frame\">\n\n    <div\n      #panElement\n      class=\"pan-element\">\n\n        <div\n          #zoomElement\n          class=\"zoom-element\">\n\n            <ng-content></ng-content>\n\n        </div>\n\n    </div>\n\n</div>\n\n<div\n  #panzoomOverlay\n  class=\"pan-zoom-overlay\">\n</div>",
                    styles: [".pan-zoom-frame{position:relative;width:100%;height:100%;overflow:hidden}.pan-element,.pan-zoom-overlay{position:absolute;left:0;top:0}.pan-zoom-overlay{display:none;right:0;bottom:0;opacity:0;pointer-events:none}"]
                },] }
    ];
    PanZoomComponent.ctorParameters = function () { return [
        { type: core.NgZone }
    ]; };
    PanZoomComponent.propDecorators = {
        frameElementRef: [{ type: core.ViewChild, args: ['frameElement', { static: true },] }],
        panElementRef: [{ type: core.ViewChild, args: ['panElement', { static: true },] }],
        zoomElementRef: [{ type: core.ViewChild, args: ['zoomElement', { static: true },] }],
        panzoomOverlayRef: [{ type: core.ViewChild, args: ['panzoomOverlay', { static: true },] }],
        config: [{ type: core.Input }]
    };

    var NgxPanZoomModule = /** @class */ (function () {
        function NgxPanZoomModule() {
        }
        return NgxPanZoomModule;
    }());
    NgxPanZoomModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [],
                    declarations: [PanZoomComponent],
                    providers: [],
                    exports: [PanZoomComponent]
                },] }
    ];

    var PanZoomConfig = /** @class */ (function () {
        function PanZoomConfig(options) {
            this.zoomLevels = 5;
            this.neutralZoomLevel = 2;
            this.scalePerZoomLevel = 2.0;
            this.initialZoomLevel = this.neutralZoomLevel;
            this.friction = 10.0;
            this.haltSpeed = 100.0;
            this.initialPanX = 0;
            this.initialPanY = 0;
            this.keepInBounds = false;
            this.keepInBoundsDragPullback = 0.7;
            this.keepInBoundsRestoreForce = 0.5;
            this.panOnClickDrag = true;
            this.dragMouseButton = 'left'; // left, middle, right
            this.zoomButtonIncrement = 1.0;
            this.zoomOnDoubleClick = true;
            this.zoomOnMouseWheel = true;
            this.invertMouseWheel = false;
            this.zoomStepDuration = 0.2;
            this.zoomToFitZoomLevelFactor = 0.95;
            this.freeMouseWheel = true;
            this.freeMouseWheelFactor = 0.08;
            this.modelChanged = new rxjs.BehaviorSubject({
                isPanning: null,
                zoomLevel: null,
                pan: {
                    x: null,
                    y: null
                }
            });
            this.api = new rxjs.BehaviorSubject({
                model: null,
                config: null,
                changeZoomLevel: null,
                zoomIn: null,
                zoomOut: null,
                zoomToFit: null,
                getViewPosition: null,
                getModelPosition: null,
                resetView: null,
                panToPoint: null,
                panDelta: null,
                panDeltaPercent: null,
                panDeltaAbsolute: null,
                centerContent: null,
                centerX: null,
                centerY: null,
                centerTopLeft: null,
                centerTopRight: null,
                centerBottomLeft: null,
                centerBottomRight: null,
                updateContentDimensions: null,
                detectContentDimensions: null
            });
            this.acceleratePan = true;
            this.dynamicContentDimensions = false;
            if (options === undefined) {
                return;
            }
            if ('zoomLevels' in options) {
                this.zoomLevels = options.zoomLevels;
            }
            if ('neutralZoomLevel' in options) {
                this.neutralZoomLevel = options.neutralZoomLevel;
            }
            if ('scalePerZoomLevel' in options) {
                this.scalePerZoomLevel = options.scalePerZoomLevel;
            }
            if ('initialZoomLevel' in options) {
                this.initialZoomLevel = options.initialZoomLevel;
            }
            if ('friction' in options) {
                this.friction = options.friction;
            }
            if ('haltSpeed' in options) {
                this.haltSpeed = options.haltSpeed;
            }
            if ('initialPanX' in options) {
                this.initialPanX = options.initialPanX;
            }
            if ('initialPanY' in options) {
                this.initialPanY = options.initialPanY;
            }
            if ('initialZoomToFit' in options) {
                this.initialZoomToFit = options.initialZoomToFit;
            }
            if ('keepInBounds' in options) {
                this.keepInBounds = options.keepInBounds;
            }
            if ('keepInBoundsDragPullback' in options) {
                this.keepInBoundsDragPullback = options.keepInBoundsDragPullback;
            }
            if ('keepInBoundsRestoreForce' in options) {
                this.keepInBoundsRestoreForce = options.keepInBoundsRestoreForce;
            }
            if ('panOnClickDrag' in options) {
                this.panOnClickDrag = options.panOnClickDrag;
            }
            if ('dragMouseButton' in options) {
                this.dragMouseButton = options.dragMouseButton;
            }
            if ('zoomButtonIncrement' in options) {
                this.zoomButtonIncrement = options.zoomButtonIncrement;
            }
            if ('zoomOnDoubleClick' in options) {
                this.zoomOnDoubleClick = options.zoomOnDoubleClick;
            }
            if ('zoomOnMouseWheel' in options) {
                this.zoomOnMouseWheel = options.zoomOnMouseWheel;
            }
            if ('invertMouseWheel' in options) {
                this.invertMouseWheel = options.invertMouseWheel;
            }
            if ('zoomStepDuration' in options) {
                this.zoomStepDuration = options.zoomStepDuration;
            }
            if ('zoomToFitZoomLevelFactor' in options) {
                this.zoomToFitZoomLevelFactor = options.zoomToFitZoomLevelFactor;
            }
            if ('freeMouseWheel' in options) {
                this.freeMouseWheel = options.freeMouseWheel;
            }
            if ('freeMouseWheelFactor' in options) {
                this.freeMouseWheelFactor = options.freeMouseWheelFactor;
            }
            if (this.keepInBounds && this.neutralZoomLevel !== 0) {
                console.warn('You have set keepInBounds to true and neutralZoomLevel to ' + this.neutralZoomLevel + '. Be aware that the zoom level cannot go below ' + this.neutralZoomLevel);
            }
            if ('noDragFromElementClass' in options) {
                this.noDragFromElementClass = options.noDragFromElementClass;
            }
            if ('acceleratePan' in options) {
                this.acceleratePan = options.acceleratePan;
            }
            if ('dynamicContentDimensions' in options) {
                this.dynamicContentDimensions = options.dynamicContentDimensions;
            }
        }
        return PanZoomConfig;
    }());

    /*
     * Public API Surface of ngx-panzoom
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NgxPanZoomModule = NgxPanZoomModule;
    exports.PanZoomComponent = PanZoomComponent;
    exports.PanZoomConfig = PanZoomConfig;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
