// @angular/compiler-cli/linker/src/file_linker/linker_environment.mjs
import { SourceFileLoader } from '../../../src/ngtsc/sourcemaps';
import { DEFAULT_LINKER_OPTIONS } from './linker_options';
import { Translator } from './translator';
export class LinkerEnvironment {
    constructor(fileSystem, logger, host, factory, options) {
        this.fileSystem = fileSystem;
        this.logger = logger;
        this.host = host;
        this.factory = factory;
        this.options = options;
        this.translator = new Translator(this.factory);
        this.sourceFileLoader = this.options.sourceMapping ? new SourceFileLoader(this.fileSystem, this.logger, {}) : null;
    }
    static create(fileSystem, logger, host, factory, options) {
        var _a, _b, _c;
        return new LinkerEnvironment(fileSystem, logger, host, factory, {
            sourceMapping: (_a = options.sourceMapping) !== null && _a !== void 0 ? _a : DEFAULT_LINKER_OPTIONS.sourceMapping,
            linkerJitMode: (_b = options.linkerJitMode) !== null && _b !== void 0 ? _b : DEFAULT_LINKER_OPTIONS.linkerJitMode,
            unknownDeclarationVersionHandling: (_c = options.unknownDeclarationVersionHandling) !== null && _c !== void 0 ? _c : DEFAULT_LINKER_OPTIONS.unknownDeclarationVersionHandling,
        });
    }
}
