// @angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_native.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/// <reference types="node" />
import * as os from 'os';
import { NodeJSFileSystem } from '../../src/node_js_file_system';
import { MockFileSystem } from './mock_file_system';
const isWindows = os.platform() === 'win32';
export class MockFileSystemNative extends MockFileSystem {
    constructor(cwd = '/') {
        super(undefined, cwd);
    }
    // Delegate to the real NodeJSFileSystem for these path related methods
    resolve(...paths) {
        return NodeJSFileSystem.prototype.resolve.call(this, this.pwd(), ...paths);
    }
    dirname(file) {
        return NodeJSFileSystem.prototype.dirname.call(this, file);
    }
    join(basePath, ...paths) {
        return NodeJSFileSystem.prototype.join.call(this, basePath, ...paths);
    }
    relative(from, to) {
        return NodeJSFileSystem.prototype.relative.call(this, from, to);
    }
    basename(filePath, extension) {
        return NodeJSFileSystem.prototype.basename.call(this, filePath, extension);
    }
    isCaseSensitive() {
        return NodeJSFileSystem.prototype.isCaseSensitive.call(this);
    }
    isRooted(path) {
        return NodeJSFileSystem.prototype.isRooted.call(this, path);
    }
    isRoot(path) {
        return NodeJSFileSystem.prototype.isRoot.call(this, path);
    }
    normalize(path) {
        // When running in Windows, absolute paths are normalized to always include a drive letter. This
        // ensures that rooted posix paths used in tests will be normalized to real Windows paths, i.e.
        // including a drive letter. Note that the same normalization is done in emulated Windows mode
        // (see `MockFileSystemWindows`) so that the behavior is identical between native Windows and
        // emulated Windows mode.
        if (isWindows) {
            path = path.replace(/^[\/\\]/i, 'C:/');
        }
        return NodeJSFileSystem.prototype.normalize.call(this, path);
    }
    splitPath(path) {
        return path.split(/[\\\/]/);
    }
}
