// @angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_posix.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/// <reference types="node" />
import * as p from 'path';
import { MockFileSystem } from './mock_file_system';
export class MockFileSystemPosix extends MockFileSystem {
    resolve(...paths) {
        const resolved = p.posix.resolve(this.pwd(), ...paths);
        return this.normalize(resolved);
    }
    dirname(file) {
        return this.normalize(p.posix.dirname(file));
    }
    join(basePath, ...paths) {
        return this.normalize(p.posix.join(basePath, ...paths));
    }
    relative(from, to) {
        return this.normalize(p.posix.relative(from, to));
    }
    basename(filePath, extension) {
        return p.posix.basename(filePath, extension);
    }
    isRooted(path) {
        return path.startsWith('/');
    }
    splitPath(path) {
        return path.split('/');
    }
    normalize(path) {
        return path.replace(/^[a-z]:\//i, '/').replace(/\\/g, '/');
    }
}
