// @angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_windows.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/// <reference types="node" />
import * as p from 'path';
import { MockFileSystem } from './mock_file_system';
export class MockFileSystemWindows extends MockFileSystem {
    resolve(...paths) {
        const resolved = p.win32.resolve(this.pwd(), ...paths);
        return this.normalize(resolved);
    }
    dirname(path) {
        return this.normalize(p.win32.dirname(path));
    }
    join(basePath, ...paths) {
        return this.normalize(p.win32.join(basePath, ...paths));
    }
    relative(from, to) {
        return this.normalize(p.win32.relative(from, to));
    }
    basename(filePath, extension) {
        return p.win32.basename(filePath, extension);
    }
    isRooted(path) {
        return /^([A-Z]:)?([\\\/]|$)/i.test(path);
    }
    splitPath(path) {
        return path.split(/[\\\/]/);
    }
    normalize(path) {
        return path.replace(/^[\/\\]/i, 'C:/').replace(/\\/g, '/');
    }
}
