// @angular-eslint/schematics/dist/migrations/update-3-0-0/update-3-0-0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const tasks_1 = require("@angular-devkit/schematics/tasks");
const utils_1 = require("../../utils");
const updatedAngularESLintVersion = '^3.0.0';
function updateIfExists(deps, depName, updatedVersion) {
    if (deps === null || deps === void 0 ? void 0 : deps[depName]) {
        deps[depName] = updatedVersion;
    }
}
function updateRelevantDependencies(host, context) {
    return utils_1.updateJsonInTree('package.json', (json) => {
        /**
         * @angular-eslint
         */
        updateIfExists(json.devDependencies, '@angular-eslint/builder', updatedAngularESLintVersion);
        updateIfExists(json.devDependencies, '@angular-eslint/eslint-plugin', updatedAngularESLintVersion);
        updateIfExists(json.devDependencies, '@angular-eslint/eslint-plugin-template', updatedAngularESLintVersion);
        updateIfExists(json.devDependencies, '@angular-eslint/template-parser', updatedAngularESLintVersion);
        context.addTask(new tasks_1.NodePackageInstallTask());
        return json;
    })(host, context);
}
function addRecommendedExtraExtendsWhereApplicable(config) {
    // Convert extends to array if applicable
    if (typeof config.extends === 'string' &&
        config.extends === 'plugin:@angular-eslint/recommended') {
        config.extends = [config.extends];
    }
    if (Array.isArray(config.extends) &&
        config.extends.includes('plugin:@angular-eslint/recommended')) {
        config.extends.push('plugin:@angular-eslint/recommended--extra');
    }
    if (config.overrides) {
        for (const override of config.overrides) {
            if (typeof override.extends === 'string' &&
                override.extends === 'plugin:@angular-eslint/recommended') {
                override.extends = [override.extends];
            }
            if (Array.isArray(override.extends) &&
                override.extends.includes('plugin:@angular-eslint/recommended')) {
                override.extends.push('plugin:@angular-eslint/recommended--extra');
            }
        }
    }
}
function applyRecommendedExtraExtends() {
    return schematics_1.chain([
        utils_1.visitNotIgnoredFiles((filePath) => {
            if (!filePath.endsWith('.eslintrc.json')) {
                return;
            }
            return utils_1.updateJsonInTree(filePath.toString(), (json) => {
                addRecommendedExtraExtendsWhereApplicable(json);
                return json;
            });
        }),
    ]);
}
function removeNegativeValuesFromComponentMaxInlineDeclarations(rule) {
    if (!Array.isArray(rule) || rule.length !== 2)
        return;
    const [, currentSchema] = rule;
    rule[1] = Object.entries(currentSchema).reduce((accumulator, [key, value]) => {
        return Number(value) < 0 ? accumulator : Object.assign(Object.assign({}, accumulator), { [key]: value });
    }, {});
}
function updateComponentMaxInlineDeclarationsSchema({ overrides, rules, }) {
    var _a;
    removeNegativeValuesFromComponentMaxInlineDeclarations(rules === null || rules === void 0 ? void 0 : rules['@angular-eslint/component-max-inline-declarations']);
    for (const override of overrides !== null && overrides !== void 0 ? overrides : []) {
        removeNegativeValuesFromComponentMaxInlineDeclarations((_a = override.rules) === null || _a === void 0 ? void 0 : _a['@angular-eslint/component-max-inline-declarations']);
    }
}
function updateComponentMaxInlineDeclarations() {
    return schematics_1.chain([
        utils_1.visitNotIgnoredFiles((filePath) => {
            if (!filePath.endsWith('.eslintrc.json')) {
                return;
            }
            return utils_1.updateJsonInTree(filePath.toString(), (json) => {
                updateComponentMaxInlineDeclarationsSchema(json);
                return json;
            });
        }),
    ]);
}
function default_1() {
    return schematics_1.chain([
        updateRelevantDependencies,
        applyRecommendedExtraExtends,
        updateComponentMaxInlineDeclarations,
    ]);
}
exports.default = default_1;
