// @angular/compiler-cli/ngcc/src/rendering/dts_renderer.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import MagicString from 'magic-string';
import * as ts from 'typescript';
import { ImportManager, translateType } from '../../../src/ngtsc/translator';
import { IMPORT_PREFIX } from '../constants';
import { renderSourceAndMap } from './source_maps';
import { getImportRewriter } from './utils';
/**
 * A structure that captures information about what needs to be rendered
 * in a typings file.
 *
 * It is created as a result of processing the analysis passed to the renderer.
 *
 * The `renderDtsFile()` method consumes it when rendering a typings file.
 */
class DtsRenderInfo {
    constructor() {
        this.classInfo = [];
        this.moduleWithProviders = [];
        this.privateExports = [];
        this.reexports = [];
    }
}
/**
 * A base-class for rendering an `AnalyzedFile`.
 *
 * Package formats have output files that must be rendered differently. Concrete sub-classes must
 * implement the `addImports`, `addDefinitions` and `removeDecorators` abstract methods.
 */
export class DtsRenderer {
    constructor(dtsFormatter, fs, logger, host, bundle) {
        this.dtsFormatter = dtsFormatter;
        this.fs = fs;
        this.logger = logger;
        this.host = host;
        this.bundle = bundle;
    }
    renderProgram(decorationAnalyses, privateDeclarationsAnalyses, moduleWithProvidersAnalyses) {
        const renderedFiles = [];
        // Transform the .d.ts files
        if (this.bundle.dts) {
            const dtsFiles = this.getTypingsFilesToRender(decorationAnalyses, privateDeclarationsAnalyses, moduleWithProvidersAnalyses);
            // If the dts entry-point is not already there (it did not have compiled classes)
            // then add it now, to ensure it gets its extra exports rendered.
            if (!dtsFiles.has(this.bundle.dts.file)) {
                dtsFiles.set(this.bundle.dts.file, new DtsRenderInfo());
            }
            dtsFiles.forEach((renderInfo, file) => renderedFiles.push(...this.renderDtsFile(file, renderInfo)));
        }
        return renderedFiles;
    }
    renderDtsFile(dtsFile, renderInfo) {
        const outputText = new MagicString(dtsFile.text);
        const printer = ts.createPrinter();
        const importManager = new ImportManager(getImportRewriter(this.bundle.dts.r3SymbolsFile, this.bundle.isCore, false), IMPORT_PREFIX);
        renderInfo.classInfo.forEach(dtsClass => {
            const endOfClass = dtsClass.dtsDeclaration.getEnd();
            dtsClass.compilation.forEach(declaration => {
                const type = translateType(declaration.type, importManager);
                markForEmitAsSingleLine(type);
                const typeStr = printer.printNode(ts.EmitHint.Unspecified, type, dtsFile);
                const newStatement = `    static ${declaration.name}: ${typeStr};\n`;
                outputText.appendRight(endOfClass - 1, newStatement);
            });
        });
        if (renderInfo.reexports.length > 0) {
            for (const e of renderInfo.reexports) {
                const newStatement = `\nexport {${e.symbolName} as ${e.asAlias}} from '${e.fromModule}';`;
                outputText.append(newStatement);
            }
        }
        this.dtsFormatter.addModuleWithProvidersParams(outputText, renderInfo.moduleWithProviders, importManager);
        this.dtsFormatter.addExports(outputText, dtsFile.fileName, renderInfo.privateExports, importManager, dtsFile);
        this.dtsFormatter.addImports(outputText, importManager.getAllImports(dtsFile.fileName), dtsFile);
        return renderSourceAndMap(this.logger, this.fs, dtsFile, outputText);
    }
    getTypingsFilesToRender(decorationAnalyses, privateDeclarationsAnalyses, moduleWithProvidersAnalyses) {
        const dtsMap = new Map();
        // Capture the rendering info from the decoration analyses
        decorationAnalyses.forEach(compiledFile => {
            let appliedReexports = false;
            compiledFile.compiledClasses.forEach(compiledClass => {
                const dtsDeclaration = this.host.getDtsDeclaration(compiledClass.declaration);
                if (dtsDeclaration) {
                    const dtsFile = dtsDeclaration.getSourceFile();
                    const renderInfo = dtsMap.has(dtsFile) ? dtsMap.get(dtsFile) : new DtsRenderInfo();
                    renderInfo.classInfo.push({ dtsDeclaration, compilation: compiledClass.compilation });
                    // Only add re-exports if the .d.ts tree is overlayed with the .js tree, as re-exports in
                    // ngcc are only used to support deep imports into e.g. commonjs code. For a deep import
                    // to work, the typing file and JS file must be in parallel trees. This logic will detect
                    // the simplest version of this case, which is sufficient to handle most commonjs
                    // libraries.
                    if (!appliedReexports &&
                        compiledClass.declaration.getSourceFile().fileName ===
                            dtsFile.fileName.replace(/\.d\.ts$/, '.js')) {
                        renderInfo.reexports.push(...compiledFile.reexports);
                        appliedReexports = true;
                    }
                    dtsMap.set(dtsFile, renderInfo);
                }
            });
        });
        // Capture the ModuleWithProviders functions/methods that need updating
        if (moduleWithProvidersAnalyses !== null) {
            moduleWithProvidersAnalyses.forEach((moduleWithProvidersToFix, dtsFile) => {
                const renderInfo = dtsMap.has(dtsFile) ? dtsMap.get(dtsFile) : new DtsRenderInfo();
                renderInfo.moduleWithProviders = moduleWithProvidersToFix;
                dtsMap.set(dtsFile, renderInfo);
            });
        }
        // Capture the private declarations that need to be re-exported
        if (privateDeclarationsAnalyses.length) {
            privateDeclarationsAnalyses.forEach(e => {
                if (!e.dtsFrom) {
                    throw new Error(`There is no typings path for ${e.identifier} in ${e.from}.\n` +
                        `We need to add an export for this class to a .d.ts typings file because ` +
                        `Angular compiler needs to be able to reference this class in compiled code, such as templates.\n` +
                        `The simplest fix for this is to ensure that this class is exported from the package's entry-point.`);
                }
            });
            const dtsEntryPoint = this.bundle.dts.file;
            const renderInfo = dtsMap.has(dtsEntryPoint) ? dtsMap.get(dtsEntryPoint) : new DtsRenderInfo();
            renderInfo.privateExports = privateDeclarationsAnalyses;
            dtsMap.set(dtsEntryPoint, renderInfo);
        }
        return dtsMap;
    }
}
function markForEmitAsSingleLine(node) {
    ts.setEmitFlags(node, ts.EmitFlags.SingleLine);
    ts.forEachChild(node, markForEmitAsSingleLine);
}
