// @angular/compiler-cli/linker/babel/src/ast/babel_ast_factory.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as t from '@babel/types';
import { assert } from '../../../../linker';
/**
 * A Babel flavored implementation of the AstFactory.
 */
export class BabelAstFactory {
    constructor(
    /** The absolute path to the source file being compiled. */
    sourceUrl) {
        this.sourceUrl = sourceUrl;
        this.createArrayLiteral = t.arrayExpression;
        this.createBlock = t.blockStatement;
        this.createConditional = t.conditionalExpression;
        this.createExpressionStatement = t.expressionStatement;
        this.createIdentifier = t.identifier;
        this.createIfStatement = t.ifStatement;
        this.createNewExpression = t.newExpression;
        this.createParenthesizedExpression = t.parenthesizedExpression;
        this.createReturnStatement = t.returnStatement;
        this.createThrowStatement = t.throwStatement;
        this.createUnaryExpression = t.unaryExpression;
    }
    attachComments(statement, leadingComments) {
        // We must process the comments in reverse because `t.addComment()` will add new ones in front.
        for (let i = leadingComments.length - 1; i >= 0; i--) {
            const comment = leadingComments[i];
            t.addComment(statement, 'leading', comment.toString(), !comment.multiline);
        }
    }
    createAssignment(target, value) {
        assert(target, isLExpression, 'must be a left hand side expression');
        return t.assignmentExpression('=', target, value);
    }
    createBinaryExpression(leftOperand, operator, rightOperand) {
        switch (operator) {
            case '&&':
            case '||':
            case '??':
                return t.logicalExpression(operator, leftOperand, rightOperand);
            default:
                return t.binaryExpression(operator, leftOperand, rightOperand);
        }
    }
    createCallExpression(callee, args, pure) {
        const call = t.callExpression(callee, args);
        if (pure) {
            t.addComment(call, 'leading', ' @__PURE__ ', /* line */ false);
        }
        return call;
    }
    createElementAccess(expression, element) {
        return t.memberExpression(expression, element, /* computed */ true);
    }
    createFunctionDeclaration(functionName, parameters, body) {
        assert(body, t.isBlockStatement, 'a block');
        return t.functionDeclaration(t.identifier(functionName), parameters.map(param => t.identifier(param)), body);
    }
    createFunctionExpression(functionName, parameters, body) {
        assert(body, t.isBlockStatement, 'a block');
        const name = functionName !== null ? t.identifier(functionName) : null;
        return t.functionExpression(name, parameters.map(param => t.identifier(param)), body);
    }
    createLiteral(value) {
        if (typeof value === 'string') {
            return t.stringLiteral(value);
        }
        else if (typeof value === 'number') {
            return t.numericLiteral(value);
        }
        else if (typeof value === 'boolean') {
            return t.booleanLiteral(value);
        }
        else if (value === undefined) {
            return t.identifier('undefined');
        }
        else if (value === null) {
            return t.nullLiteral();
        }
        else {
            throw new Error(`Invalid literal: ${value} (${typeof value})`);
        }
    }
    createObjectLiteral(properties) {
        return t.objectExpression(properties.map(prop => {
            const key = prop.quoted ? t.stringLiteral(prop.propertyName) : t.identifier(prop.propertyName);
            return t.objectProperty(key, prop.value);
        }));
    }
    createPropertyAccess(expression, propertyName) {
        return t.memberExpression(expression, t.identifier(propertyName), /* computed */ false);
    }
    createTaggedTemplate(tag, template) {
        const elements = template.elements.map((element, i) => this.setSourceMapRange(t.templateElement(element, i === template.elements.length - 1), element.range));
        return t.taggedTemplateExpression(tag, t.templateLiteral(elements, template.expressions));
    }
    createTypeOfExpression(expression) {
        return t.unaryExpression('typeof', expression);
    }
    createVariableDeclaration(variableName, initializer, type) {
        return t.variableDeclaration(type, [t.variableDeclarator(t.identifier(variableName), initializer)]);
    }
    setSourceMapRange(node, sourceMapRange) {
        if (sourceMapRange === null) {
            return node;
        }
        node.loc = {
            // Add in the filename so that we can map to external template files.
            // Note that Babel gets confused if you specify a filename when it is the original source
            // file. This happens when the template is inline, in which case just use `undefined`.
            filename: sourceMapRange.url !== this.sourceUrl ? sourceMapRange.url : undefined,
            start: {
                line: sourceMapRange.start.line + 1,
                column: sourceMapRange.start.column,
            },
            end: {
                line: sourceMapRange.end.line + 1,
                column: sourceMapRange.end.column,
            },
        }; // Needed because the Babel typings for `loc` don't include `filename`.
        node.start = sourceMapRange.start.offset;
        node.end = sourceMapRange.end.offset;
        return node;
    }
}
function isLExpression(expr) {
    // Some LVal types are not expressions, which prevents us from using `t.isLVal()`
    // directly with `assert()`.
    return t.isLVal(expr);
}
