// @angular/compiler-cli/src/ngtsc/scope/src/typecheck.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { CssSelector, SelectorMatcher } from '@angular/compiler';
import * as ts from 'typescript';
import { flattenInheritedDirectiveMetadata } from '../../metadata';
/**
 * Computes scope information to be used in template type checking.
 */
export class TypeCheckScopeRegistry {
    constructor(scopeReader, metaReader) {
        this.scopeReader = scopeReader;
        this.metaReader = metaReader;
        /**
         * Cache of flattened directive metadata. Because flattened metadata is scope-invariant it's
         * cached individually, such that all scopes refer to the same flattened metadata.
         */
        this.flattenedDirectiveMetaCache = new Map();
        /**
         * Cache of the computed type check scope per NgModule declaration.
         */
        this.scopeCache = new Map();
    }
    /**
     * Computes the type-check scope information for the component declaration. If the NgModule
     * contains an error, then 'error' is returned. If the component is not declared in any NgModule,
     * an empty type-check scope is returned.
     */
    getTypeCheckScope(node) {
        const matcher = new SelectorMatcher();
        const directives = [];
        const pipes = new Map();
        const scope = this.scopeReader.getScopeForComponent(node);
        if (scope === null) {
            return {
                matcher,
                directives,
                pipes,
                schemas: [],
                isPoisoned: false,
            };
        }
        if (this.scopeCache.has(scope.ngModule)) {
            return this.scopeCache.get(scope.ngModule);
        }
        for (const meta of scope.compilation.directives) {
            if (meta.selector !== null) {
                const extMeta = this.getTypeCheckDirectiveMetadata(meta.ref);
                matcher.addSelectables(CssSelector.parse(meta.selector), extMeta);
                directives.push(extMeta);
            }
        }
        for (const { name, ref } of scope.compilation.pipes) {
            if (!ts.isClassDeclaration(ref.node)) {
                throw new Error(`Unexpected non-class declaration ${ts.SyntaxKind[ref.node.kind]} for pipe ${ref.debugName}`);
            }
            pipes.set(name, ref);
        }
        const typeCheckScope = {
            matcher,
            directives,
            pipes,
            schemas: scope.schemas,
            isPoisoned: scope.compilation.isPoisoned || scope.exported.isPoisoned,
        };
        this.scopeCache.set(scope.ngModule, typeCheckScope);
        return typeCheckScope;
    }
    getTypeCheckDirectiveMetadata(ref) {
        const clazz = ref.node;
        if (this.flattenedDirectiveMetaCache.has(clazz)) {
            return this.flattenedDirectiveMetaCache.get(clazz);
        }
        const meta = flattenInheritedDirectiveMetadata(this.metaReader, ref);
        this.flattenedDirectiveMetaCache.set(clazz, meta);
        return meta;
    }
}
