// @angular/compiler-cli/linker/src/file_linker/emit_scopes/iife_emit_scope.mjs
import { EmitScope } from './emit_scope';
/**
 * This class is a specialization of the `EmitScope` class that is designed for the situation where
 * there is no clear shared scope for constant statements. In this case they are bundled with the
 * translated definition inside an IIFE.
 */
export class IifeEmitScope extends EmitScope {
    constructor(ngImport, translator, factory) {
        super(ngImport, translator);
        this.factory = factory;
    }
    /**
     * Translate the given Output AST definition expression into a generic `TExpression`.
     *
     * Wraps the output from `EmitScope.translateDefinition()` and `EmitScope.getConstantStatements()`
     * in an IIFE.
     */
    translateDefinition(definition) {
        const constantStatements = super.getConstantStatements();
        const returnStatement = this.factory.createReturnStatement(super.translateDefinition(definition));
        const body = this.factory.createBlock([...constantStatements, returnStatement]);
        const fn = this.factory.createFunctionExpression(/* name */ null, /* args */ [], body);
        return this.factory.createCallExpression(fn, /* args */ [], /* pure */ false);
    }
    /**
     * It is not valid to call this method, since there will be no shared constant statements - they
     * are already emitted in the IIFE alongside the translated definition.
     */
    getConstantStatements() {
        throw new Error('BUG - IifeEmitScope should not expose any constant statements');
    }
}
