// @angular/compiler-cli/ngcc/src/packages/entry_point.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as ts from 'typescript';
import { parseStatementForUmdModule } from '../host/umd_host';
import { resolveFileWithPostfixes } from '../utils';
// We need to keep the elements of this const and the `EntryPointJsonProperty` type in sync.
export const SUPPORTED_FORMAT_PROPERTIES = ['fesm2015', 'fesm5', 'es2015', 'esm2015', 'esm5', 'main', 'module', 'browser'];
/**
 * The path does not represent an entry-point, i.e. there is no package.json at the path and there
 * is no config to force an entry-point.
 */
export const NO_ENTRY_POINT = 'no-entry-point';
/**
 * The path represents an entry-point that is `ignored` by an ngcc config.
 */
export const IGNORED_ENTRY_POINT = 'ignored-entry-point';
/**
 * The path has a package.json, but it is not a valid entry-point for ngcc processing.
 */
export const INCOMPATIBLE_ENTRY_POINT = 'incompatible-entry-point';
/**
 * Try to create an entry-point from the given paths and properties.
 *
 * @param packagePath the absolute path to the containing npm package
 * @param entryPointPath the absolute path to the potential entry-point.
 * @returns
 * - An entry-point if it is valid and not ignored.
 * - `NO_ENTRY_POINT` when there is no package.json at the path and there is no config to force an
 *   entry-point,
 * - `IGNORED_ENTRY_POINT` when the entry-point is ignored by an ngcc config.
 * - `INCOMPATIBLE_ENTRY_POINT` when there is a package.json but it is not a valid Angular compiled
 *   entry-point.
 */
export function getEntryPointInfo(fs, config, logger, packagePath, entryPointPath) {
    const packagePackageJsonPath = fs.resolve(packagePath, 'package.json');
    const entryPointPackageJsonPath = fs.resolve(entryPointPath, 'package.json');
    const loadedPackagePackageJson = loadPackageJson(fs, packagePackageJsonPath);
    const loadedEntryPointPackageJson = (packagePackageJsonPath === entryPointPackageJsonPath) ?
        loadedPackagePackageJson :
        loadPackageJson(fs, entryPointPackageJsonPath);
    const { packageName, packageVersion } = getPackageNameAndVersion(fs, packagePath, loadedPackagePackageJson, loadedEntryPointPackageJson);
    const packageConfig = config.getPackageConfig(packageName, packagePath, packageVersion);
    const entryPointConfig = packageConfig.entryPoints.get(entryPointPath);
    let entryPointPackageJson;
    if (entryPointConfig === undefined) {
        if (!fs.exists(entryPointPackageJsonPath)) {
            // No `package.json` and no config.
            return NO_ENTRY_POINT;
        }
        else if (loadedEntryPointPackageJson === null) {
            // `package.json` exists but could not be parsed and there is no redeeming config.
            logger.warn(`Failed to read entry point info from invalid 'package.json' file: ${entryPointPackageJsonPath}`);
            return INCOMPATIBLE_ENTRY_POINT;
        }
        else {
            entryPointPackageJson = loadedEntryPointPackageJson;
        }
    }
    else if (entryPointConfig.ignore === true) {
        // Explicitly ignored entry-point.
        return IGNORED_ENTRY_POINT;
    }
    else {
        entryPointPackageJson = mergeConfigAndPackageJson(fs, loadedEntryPointPackageJson, entryPointConfig, packagePath, entryPointPath);
    }
    const typings = entryPointPackageJson.typings || entryPointPackageJson.types ||
        guessTypingsFromPackageJson(fs, entryPointPath, entryPointPackageJson);
    if (typeof typings !== 'string') {
        // Missing the required `typings` property
        return INCOMPATIBLE_ENTRY_POINT;
    }
    // An entry-point is assumed to be compiled by Angular if there is either:
    // * a `metadata.json` file next to the typings entry-point
    // * a custom config for this entry-point
    const metadataPath = fs.resolve(entryPointPath, typings.replace(/\.d\.ts$/, '') + '.metadata.json');
    const compiledByAngular = entryPointConfig !== undefined || fs.exists(metadataPath);
    const entryPointInfo = {
        name: entryPointPackageJson.name,
        path: entryPointPath,
        packageName,
        packagePath,
        packageJson: entryPointPackageJson,
        typings: fs.resolve(entryPointPath, typings),
        compiledByAngular,
        ignoreMissingDependencies: entryPointConfig !== undefined ? !!entryPointConfig.ignoreMissingDependencies : false,
        generateDeepReexports: entryPointConfig !== undefined ? !!entryPointConfig.generateDeepReexports : false,
    };
    return entryPointInfo;
}
export function isEntryPoint(result) {
    return result !== NO_ENTRY_POINT && result !== INCOMPATIBLE_ENTRY_POINT &&
        result !== IGNORED_ENTRY_POINT;
}
/**
 * Convert a package.json property into an entry-point format.
 *
 * @param property The property to convert to a format.
 * @returns An entry-point format or `undefined` if none match the given property.
 */
export function getEntryPointFormat(fs, entryPoint, property) {
    switch (property) {
        case 'fesm2015':
            return 'esm2015';
        case 'fesm5':
            return 'esm5';
        case 'es2015':
            return 'esm2015';
        case 'esm2015':
            return 'esm2015';
        case 'esm5':
            return 'esm5';
        case 'browser':
            const browserFile = entryPoint.packageJson['browser'];
            if (typeof browserFile !== 'string') {
                return undefined;
            }
            return sniffModuleFormat(fs, fs.join(entryPoint.path, browserFile));
        case 'main':
            const mainFile = entryPoint.packageJson['main'];
            if (mainFile === undefined) {
                return undefined;
            }
            return sniffModuleFormat(fs, fs.join(entryPoint.path, mainFile));
        case 'module':
            const moduleFilePath = entryPoint.packageJson['module'];
            // As of version 10, the `module` property in `package.json` should point to
            // the ESM2015 format output as per Angular Package format specification. This
            // means that the `module` property captures multiple formats, as old libraries
            // built with the old APF can still be processed. We detect the format by checking
            // the paths that should be used as per APF specification.
            if (typeof moduleFilePath === 'string' && moduleFilePath.includes('esm2015')) {
                return `esm2015`;
            }
            return 'esm5';
        default:
            return undefined;
    }
}
/**
 * Parse the JSON from a `package.json` file.
 * @param packageJsonPath the absolute path to the `package.json` file.
 * @returns JSON from the `package.json` file if it is valid, `null` otherwise.
 */
function loadPackageJson(fs, packageJsonPath) {
    try {
        return JSON.parse(fs.readFile(packageJsonPath));
    }
    catch (_a) {
        return null;
    }
}
function sniffModuleFormat(fs, sourceFilePath) {
    const resolvedPath = resolveFileWithPostfixes(fs, sourceFilePath, ['', '.js', '/index.js']);
    if (resolvedPath === null) {
        return undefined;
    }
    const sourceFile = ts.createSourceFile(sourceFilePath, fs.readFile(resolvedPath), ts.ScriptTarget.ES5);
    if (sourceFile.statements.length === 0) {
        return undefined;
    }
    if (ts.isExternalModule(sourceFile)) {
        return 'esm5';
    }
    else if (parseStatementForUmdModule(sourceFile.statements[0]) !== null) {
        return 'umd';
    }
    else {
        return 'commonjs';
    }
}
function mergeConfigAndPackageJson(fs, entryPointPackageJson, entryPointConfig, packagePath, entryPointPath) {
    if (entryPointPackageJson !== null) {
        return Object.assign(Object.assign({}, entryPointPackageJson), entryPointConfig.override);
    }
    else {
        const name = `${fs.basename(packagePath)}/${fs.relative(packagePath, entryPointPath)}`;
        return Object.assign({ name }, entryPointConfig.override);
    }
}
function guessTypingsFromPackageJson(fs, entryPointPath, entryPointPackageJson) {
    for (const prop of SUPPORTED_FORMAT_PROPERTIES) {
        const field = entryPointPackageJson[prop];
        if (typeof field !== 'string') {
            // Some crazy packages have things like arrays in these fields!
            continue;
        }
        const relativeTypingsPath = field.replace(/\.js$/, '.d.ts');
        const typingsPath = fs.resolve(entryPointPath, relativeTypingsPath);
        if (fs.exists(typingsPath)) {
            return typingsPath;
        }
    }
    return null;
}
/**
 * Find or infer the name and version of a package.
 *
 * - The name is computed based on the `name` property of the package's or the entry-point's
 *   `package.json` file (if available) or inferred from the package's path.
 * - The version is read off of the `version` property of the package's `package.json` file (if
 *   available).
 *
 * @param fs The file-system to use for processing `packagePath`.
 * @param packagePath the absolute path to the package.
 * @param packagePackageJson the parsed `package.json` of the package (if available).
 * @param entryPointPackageJson the parsed `package.json` of an entry-point (if available).
 * @returns the computed name and version of the package.
 */
function getPackageNameAndVersion(fs, packagePath, packagePackageJson, entryPointPackageJson) {
    var _a;
    let packageName;
    if (packagePackageJson !== null) {
        // We have a valid `package.json` for the package: Get the package name from that.
        packageName = packagePackageJson.name;
    }
    else if (entryPointPackageJson !== null) {
        // We have a valid `package.json` for the entry-point: Get the package name from that.
        // This might be a secondary entry-point, so make sure we only keep the main package's name
        // (e.g. only keep `@angular/common` from `@angular/common/http`).
        packageName = /^(?:@[^/]+\/)?[^/]*/.exec(entryPointPackageJson.name)[0];
    }
    else {
        // We don't have a valid `package.json`: Infer the package name from the package's path.
        const lastSegment = fs.basename(packagePath);
        const secondLastSegment = fs.basename(fs.dirname(packagePath));
        packageName =
            secondLastSegment.startsWith('@') ? `${secondLastSegment}/${lastSegment}` : lastSegment;
    }
    return {
        packageName,
        packageVersion: (_a = packagePackageJson === null || packagePackageJson === void 0 ? void 0 : packagePackageJson.version) !== null && _a !== void 0 ? _a : null,
    };
}
