// @angular/compiler-cli/src/diagnostics/translate_diagnostics.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as ts from 'typescript';
import { DEFAULT_ERROR_CODE, SOURCE } from '../transformers/api';
import { GENERATED_FILES } from '../transformers/util';
export function translateDiagnostics(host, untranslatedDiagnostics) {
    const ts = [];
    const ng = [];
    untranslatedDiagnostics.forEach((diagnostic) => {
        if (diagnostic.file && diagnostic.start && GENERATED_FILES.test(diagnostic.file.fileName)) {
            // We need to filter out diagnostics about unused functions as
            // they are in fact referenced by nobody and only serve to surface
            // type check errors.
            if (diagnostic.code === /* ... is declared but never used */ 6133) {
                return;
            }
            const span = sourceSpanOf(host, diagnostic.file, diagnostic.start);
            if (span) {
                const fileName = span.start.file.url;
                ng.push({
                    messageText: diagnosticMessageToString(diagnostic.messageText),
                    category: diagnostic.category,
                    span,
                    source: SOURCE,
                    code: DEFAULT_ERROR_CODE
                });
            }
        }
        else {
            ts.push(diagnostic);
        }
    });
    return { ts, ng };
}
function sourceSpanOf(host, source, start) {
    const { line, character } = ts.getLineAndCharacterOfPosition(source, start);
    return host.parseSourceSpanOf(source.fileName, line, character);
}
function diagnosticMessageToString(message) {
    return ts.flattenDiagnosticMessageText(message, '\n');
}
