// @angular/compiler-cli/src/ngtsc/routing/src/route.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
export class RouterEntryPoint {
}
class RouterEntryPointImpl {
    constructor(filePath, moduleName) {
        this.filePath = filePath;
        this.moduleName = moduleName;
    }
    get name() {
        return this.moduleName;
    }
    // For debugging purposes.
    toString() {
        return `RouterEntryPoint(name: ${this.name}, filePath: ${this.filePath})`;
    }
}
export class RouterEntryPointManager {
    constructor(moduleResolver) {
        this.moduleResolver = moduleResolver;
        this.map = new Map();
    }
    resolveLoadChildrenIdentifier(loadChildrenIdentifier, context) {
        const [relativeFile, moduleName] = loadChildrenIdentifier.split('#');
        if (moduleName === undefined) {
            return null;
        }
        const resolvedSf = this.moduleResolver.resolveModule(relativeFile, context.fileName);
        if (resolvedSf === null) {
            return null;
        }
        return this.fromNgModule(resolvedSf, moduleName);
    }
    fromNgModule(sf, moduleName) {
        const key = entryPointKeyFor(sf.fileName, moduleName);
        if (!this.map.has(key)) {
            this.map.set(key, new RouterEntryPointImpl(sf.fileName, moduleName));
        }
        return this.map.get(key);
    }
}
export function entryPointKeyFor(filePath, moduleName) {
    // Drop the extension to be compatible with how cli calls `listLazyRoutes(entryRoute)`.
    return `${filePath.replace(/\.tsx?$/i, '')}#${moduleName}`;
}
