// @angular/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.mjs
import { getBasePaths, trackDuration } from './utils';
/**
 * An EntryPointFinder that searches for all entry-points that can be found given a `basePath` and
 * `pathMappings`.
 */
export class DirectoryWalkerEntryPointFinder {
    constructor(logger, resolver, entryPointCollector, entryPointManifest, sourceDirectory, pathMappings) {
        this.logger = logger;
        this.resolver = resolver;
        this.entryPointCollector = entryPointCollector;
        this.entryPointManifest = entryPointManifest;
        this.sourceDirectory = sourceDirectory;
        this.pathMappings = pathMappings;
        this.basePaths = getBasePaths(this.logger, this.sourceDirectory, this.pathMappings);
    }
    /**
     * Search the `sourceDirectory`, and sub-directories, using `pathMappings` as necessary, to find
     * all package entry-points.
     */
    findEntryPoints() {
        const unsortedEntryPoints = [];
        for (const basePath of this.basePaths) {
            const entryPoints = this.entryPointManifest.readEntryPointsUsingManifest(basePath) ||
                this.walkBasePathForPackages(basePath);
            entryPoints.forEach(e => unsortedEntryPoints.push(e));
        }
        return this.resolver.sortEntryPointsByDependency(unsortedEntryPoints);
    }
    /**
     * Search the `basePath` for possible Angular packages and entry-points.
     *
     * @param basePath The path at which to start the search.
     * @returns an array of `EntryPoint`s that were found within `basePath`.
     */
    walkBasePathForPackages(basePath) {
        this.logger.debug(`No manifest found for ${basePath} so walking the directories for entry-points.`);
        const entryPoints = trackDuration(() => this.entryPointCollector.walkDirectoryForPackages(basePath), duration => this.logger.debug(`Walking ${basePath} for entry-points took ${duration}s.`));
        this.entryPointManifest.writeEntryPointManifest(basePath, entryPoints);
        return entryPoints;
    }
}
